CFE-748: Improve CoreDNS Integration with EgressFirewall (#1335)
openshift/enhancements: enhancements/dns/coredns-egressfirewall-integration.md

Reviewer: What will be done for DNS names that belong to multiple IP addresses?

Author: All of them or else the egress firewall rules won't be implemented correctly.

Reviewer: Does the egress firewall cross-reference the dns name once it adds ip addresses?

Author: Yes. It uses the TTL of the DNS record for a re-resolution: https://github.com/openshift/enhancements/pull/1335/files#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R37-R39

Author: You can also refer the docs: https://docs.openshift.com/container-platform/4.12/networking/ovn_kubernetes_network_provider/configuring-egress-firewall-ovn.html#domain-name-server-resolution_configuring-egress-firewall-ovn

Reviewer: You may want to exclude this case from your goals, e.g. when one, non-wildcard, DNS name resolves to multiple IP addresses.  Any of those IP addresses can change over time and get out of sync with the firewall.  Since the DNS name is not wildcard, it wouldn't be included in this feature.

Author: This feature is not only for wildcard DNS. It'll also improve the integration between CoreDNS and Egress Firewall to support the scenario where the IP addresses corresponding to a DNS name dynamically changes.

As explained by @danwinship here (https://github.com/openshift/enhancements/pull/1335#discussion_r1149397605), the feature's first goal is to "change the way that all egress firewall DNS works" for the problem described in [RFE-1355](https://issues.redhat.com/browse/RFE-1355). There was an existing RFE ([RFE-1384](https://issues.redhat.com/browse/RFE-1384)) for the support of wildcard DNS and to support wildcard DNS we need to "change the way that all egress firewall DNS works" which was already requested in RFE-1355. Thus both are combined here.

So these feature will solve the following issues:

- IP addresses corresponding to a regular DNS name going out of sync with the egress firewall (RFE-1355)
- Support wildcard DNS names in egress firewall (RFE-1384)

Reviewer: I still don't agree that combining the two features into one enhancement proposal is wise.  Taking these separately would have made this a more seamless review and supported agile and iterative development.

Reviewer: What are the changes for the EgressNetworkPolicy?

Author: > What are the changes for the EgressNetworkPolicy?

@Deepthidharwar clarified that the feature is only for OVN-K ([slack thread](https://redhat-internal.slack.com/archives/CCH60A77E/p1674134213288299))

As mentioned earlier, only DNS allow rules will be supported (https://github.com/openshift/enhancements/pull/1335#discussion_r1101797917, https://github.com/openshift/enhancements/pull/1335#discussion_r1150584471). So the deny rule will be for CIDR `0.0.0.0/0` and DNS rules will only be added for allowing DNS names or subdomains.

Author: > the deny rule will be for CIDR `0.0.0.0/0` and DNS rules will only be added for allowing DNS names or subdomains.

I will add this to the customer use case.

master then the pod will be incorrectly allowed/denied access to the DNS name. The integration between
CoreDNS and EgressFirewall needs to be improved to avoid such a scenario.

If an administrator wants to specifically allow/deny access to all subdomains then currently the
administrator has to add all subdomains in the EgressFirewall rules. This becomes difficult when
subdomains are dynamically added/removed as each one has to be added/removed individually from the
EgressFirewall rules. Currently, wildcard DNS names are not supported in EgressFirewall. However,
even if the support is added to EgressFirewall, the integration between CoreDNS and EgressFirewall
needs to be improved as wildcard DNS names cannot be directly looked up through a DNS query.

### User Stories

* As an OpenShift cluster administrator, I want to add regular DNS names to EgressFirewall rules, so that I can allow/deny
access to them even if the IPs associated with the corresponding DNS records change dynamically.
* As an OpenShift cluster administrator, I want to add wildcard DNS names to EgressFirewall rules, so that I can

> allow/deny access to all the subdomains belonging to the wildcard DNS names.

Author: Should there be validations added then that only "allow" rules can be used for DNS based firewall rules? That means creation of EgressFirewall will fail if this is not followed?

> allow/deny access to all the subdomains belonging to the wildcard DNS names.

Reviewer 2: That would be an API break at this point, unfortunately.

> allow/deny access to all the subdomains belonging to the wildcard DNS names.

Reviewer: So, is this going to change in this proposal?

> allow/deny access to all the subdomains belonging to the wildcard DNS names.

Author: Cilium also supports only allow rules (https://github.com/cilium/cilium/blob/HEAD/pkg/policy/api/egress.go#L137-L161). It makes sense to only support only allow rules for DNS names in EgressFirewall.

* As an OpenShift engineer, I want to add a new Custom Resource, so that IPs and TTLs of the DNS names which are used
in EgressFirewall rules can be tracked.
* As an OpenShift engineer, I want to add a new plugin to CoreDNS, so that DNS lookups of DNS names used in EgressFirewall
rules can be inspected and the current IPs and TTLs can be tracked in the corresponding new CR.
* As an OpenShift engineer, I want to modify Cluster DNS operator, so that CoreDNS can be deployed with the

> new plugin enabled.

Author: Ok. I will remove these user stories.

> new plugin enabled.

Reviewer: @arkadeepsen I think you were looking for new reviews on this, but I'm still seeing some comments still open, like this and others.  It makes me hesitant to do a review when I know things are going to change after I do it.

> new plugin enabled.

Author: Thanks for your reviews @candita. I was waiting for one set of review from both the SDN and NE teams before making any changes to the PR. I will wait for your complete review before making any changes to the current PR.

> new plugin enabled.

Reviewer: I won't be reviewing it any further until we come to a conclusion on wildcard dns.  I suggest splitting it into another feature, it doesn't belong combined with this feature.

### Goals

* Support update of the ``AddressSet`` referenced by the ACL rule corresponding to an EgressFirewall if IPs associated
with the corresponding DNS names change dynamically.
* Support usage of wildcard DNS names in EgressFirewall rules.
* Create CRs for each unique DNS name used in the EgressFirewall rules and use the CRs to track the current

> IPs and the corresponding TTL information.

Author: Ok. I will update the goals accordingly.

> IPs and the corresponding TTL information.

Reviewer: When will you update the goals?  If I review this and the goals change afterward, then I just might need to review again.

> IPs and the corresponding TTL information.

Author: As I mentioned in the previous comment I was waiting for one set of review from both the SDN and NE teams before making any changes to the PR.

### Non-Goals

> * Support additional DNS resolution functionality in the new CoreDNS plugin.

Author: What I meant here is the new plugin will only inspect the response of DNS resolution by other existing plugins. It will not do any DNS lookup (resolution) by itself.

> * Support allowing/denying of DNS lookups (core EgressFirewall functionality) in the new CoreDNS plugin.

Author: The CoreDNS will respond to all the DNS lookup queries. The new plugin will not stop the DNS lookup itself and respond with a NXDOMAIN if a EgressFirewall rule denies access to the specific DNS name being queried for. Thus the EgressFirewall functionality of allowing/denying will solely be the responsibility of the OVN-K master.

## Proposal

This enhancement proposes to introduce a new CoreDNS plugin (``egressfirewall``) and a new Custom Resource
(``DNSName``) to improve the integration of CoreDNS with EgressFirewall. The ``DNSName`` CR will be created
for each unique DNS name (both regular and wildcard DNS names) used in the EgressFirewall rules. This CR will
be used to store the DNS name along with the current IPs and the correspodning TTL and the next lookup time
based on the TTL. The OVN-K master will be responsible for the creation of the ``DNSName`` CRs. The ``DNSName``
CR is meant for communication between CoreDNS and OVN-K master. Thus it cannot be created or modified by an

> OpenShift cluster adminisrator.

Author: Ok. Will remove the sentence.

> OpenShift cluster adminisrator.

Reviewer 3: > you don't have to worry about what happens if a cluster-admin breaks the DNSNames

Could we add a short section to the risks/mitigations that explains what might happen if they manually start tweaking these things?

> OpenShift cluster adminisrator.

Author: Ok. I'll add risks/mitigations section.

The new plugin will inspect each DNS lookup and the corresponding response for the DNS lookup from other
plugins. If the DNS name in the query matches any ``DNSName`` CR(s) (regular or wildcard or both), then the
plugin will update the ``.status`` of the matching ``DNSName`` CR(s) with the DNS name along with the IPs and
the corresponding TTL and the next lookup time based on the TTL. The OVN-K master will watch the ``DNSName``
CRs. Whenever the IPs are updated for a ``DNSName`` CR, the OVN-K master will update the the underlying ``AdressSet``
referenced by the ACL rule(s) for the corresponding EgressFirewall(s).

OVN-K master will keep track of the TTL (or next lookup time) for each regular DNS name and send a DNS lookup
query to CoreDNS when the minimum TTL expires. However, for a wildcard DNS name a DNS lookup cannot be performed
directly on the DNS name as it will not return any IP. Thus, the lookups will be performed on the DNS names

> which are updated in the ``.status`` of the corresponding wildcard ``DNSName`` CRs.

Reviewer 4: +1, another question - if I got that right, to enable frequent IP updates, coreDNS needs feedback from the ovn-k master, to ensure ip changes are applied for egress firewall before pods start using them.
So the status may need to include a filed for that feedback from ovn-k master?

> which are updated in the ``.status`` of the corresponding wildcard ``DNSName`` CRs.

Reviewer 5: Maybe a "generation" ID in the CR that ovnkube updates? Whenever that changes CoreDNS would know that set of updates was applied.

> which are updated in the ``.status`` of the corresponding wildcard ``DNSName`` CRs.

Author: OVN-K not re-resolving the IPs whenever the TTL expires has some downsides:
1. This would mean that the additional delay introduced in the DNS lookup process for updating the new IPs in the underlying ACL will always be experienced by the client pods. If OVN-K continues to do the re-resolving of the IPs, according to the current behavior, the delay in DNS lookup will mostly be faced by OVN-K and not by the client pods. Thus the overall user experience is not degraded.
2. For wildcard DNS names, the individual subdomains are added to the ``.status`` of the ``DNSName`` CR. Now once these are added OVN-K must check whether these have become stale or not. Thus a retry counter is proposed for every DNS name which will be tracked by OVN-K. If the retry counter crosses a threshold value, then the subdomain will be removed from the ``.status`` of the ``DNSName``. If this is not done then egressfirewall will incorrectly allow certain IPs.

> which are updated in the ``.status`` of the corresponding wildcard ``DNSName`` CRs.

Author: The CoreDNS pods run as Daemonsets and they run independent of each other. So if the CoreDNS tracks the TTL and does the re-resolution, then all the CoreDNS pods would be doing the same work and there'll be chances of race condition while updating the IPs in the ``.status`` of the ``DNSName`` CRs.

> which are updated in the ``.status`` of the corresponding wildcard ``DNSName`` CRs.

Reviewer 4: How is it different from the `DNSName` creation? I mean if ovn-k master creates a new `DNSName` object, some CoreDNS pod needs to resolve it right? Which CoreDNS pod will do that?

> which are updated in the ``.status`` of the corresponding wildcard ``DNSName`` CRs.

Author: The DNS lookup request is sent to the CoreDNS service and whichever CoreDNS pod receives the DNS lookup query from OVN-K will respond to the query. If the re-resolution request is sent by the CoreDNS pods, then all of them will raise the same request as all of them are independent of each other.

> which are updated in the ``.status`` of the corresponding wildcard ``DNSName`` CRs.

Reviewer 4: oh i missed that part where ovn-k needs to send the request together with creating the `DNSName` object. so for wildcarded objects it should set random string instead of `*` to trigger resolution?
In general, sounds like a very tight integration between ovn-k and `CoreDNS` implementation, and as @danwinship said that may not match the initial goal of this feature?

> which are updated in the ``.status`` of the corresponding wildcard ``DNSName`` CRs.

Author: @npinaeva the process for wildcard DNS names is explained here: https://github.com/openshift/enhancements/pull/1335/files#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R129-R150

> which are updated in the ``.status`` of the corresponding wildcard ``DNSName`` CRs.

Author: The goal of the feature is to update the underlying ACLs for EgressFirewall first before CoreDNS responds to pods with the DNS lookup response so that pods are not incorrectly allowed/denied access to DNS names. Another goal is to support wildcard DNS names. These goals can only be supported with a better integration between EgressFirewall and CoreDNS.

Currently, the defaultTTL for re-resolving a DNS name is set to 30s in OVN-K. We can tinker with this value and set it to a higher value so that the re-resolution is done less frequently. However, as I mentioned in the [downsides](https://github.com/openshift/enhancements/pull/1335#discussion_r1107544210) of removing this existing behavior from OVN-K, both CoreDNS and EgressFirewall will not behave as expected.

> which are updated in the ``.status`` of the corresponding wildcard ``DNSName`` CRs.

Reviewer 2: Hm... I hadn't been considering the fact that there are many CoreDNS services, not just one. That makes communication via API objects much trickier...

My original idea in jira had been that CoreDNS and ovn-k would communicate via gRPC, not via the apiserver. But you still need a global view of allowed/denied IPs, because pods might end up talking to a CoreDNS server on a different node during updates when their own local DNS server is unavailable. Which means this is incompatible with OVN-IC (unless every CoreDNS plugin communicates with _every_ ovnkube-master, which seems... very complicated).

So... the plugin will have to deal with merging its response with the existing `status`... which I guess means that we _do_ need to keep the TTL data in the status, so that a plugin can distinguish "old IP that can be removed" vs "IP that another node saw and I didn't, but which is still valid".

### Workflow Description

> The workflows for DNS name addition and deletion are explained in this section.

Author: I agree. But that's how the EP template is structure. It was mentioned in the template not to remove any section. I guess moving the sections up and down also might cause a problem with the automation.

> The workflows for DNS name addition and deletion are explained in this section.

Reviewer 2: You can explain the API before the formal "API additoins" section, and then just refer back to it from there

> The workflows for DNS name addition and deletion are explained in this section.

Author: Ok. I will make the appropriate changes.

#### Addition

##### Regular DNS name
* An OpenShift cluster administrator creates an EgressFirewall resource for a Namespace and adds rule(s) containing
regular DNS name(s).
* The OVN-K master will create corrresponding `DNSName` CRs for each of the DNS names in the EgressFirewall rules, if not
already created. Each CR will be created in the ``openshift-dns`` Namespace and the Name of the CR will be same as the DNS

> name (barring any trailinng `.`). The ``.spec.isregular`` field of the CR will be set to true, even if it already exists.

Author: I used the json naming convention here which uses all lowercase. The actual fields in go are named as ``IsRegular`` and ``IsWildcard``.

> name (barring any trailinng `.`). The ``.spec.isregular`` field of the CR will be set to true, even if it already exists.

Author: Ok. I will fix the naming issues.

on the TTL. However, this update will only take place if there is a change in the exisiting IP
addresses or next time to lookup or both for the DNS name.
* The OVN-K master will watch the ``DNSName`` CRs. When the ``.status`` of a ``DNSName`` CR is updated, the OVN-K master
will update the ``AddressSet`` for the DNS name, which is linked with the ACL rule(s) for the corresponding EgressFirewall(s).
* The OVN-K master will also receive the response of the DNS lookup query for the DNS name. The OVN-K master will check
the corresponding ``DNSName`` CR's ``.status`` and if the next lookup time in the status is greater than the next lookup
time based on the received TTL, then the corresponding CR's ``.status`` will be updated. The corresponding ``AddressSet``
will also be updated.

##### Wildcard DNS name
* An OpenShift cluster administrator creates an EgressFirewall resource for a Namespace and adds rule(s) containing
wildcard DNS name(s).
* The OVN-K master will create corrresponding `DNSName` CRs for each of the wildcard DNS names in the EgressFirewall rules, if not

> already created. Each CR will be created in the ``openshift-dns`` Namespace. The Name of the CR will be set to the wildcard

Author: Ok. I will change the namespace to ``openshift-ovn-kubernetes``.

> already created. Each CR will be created in the ``openshift-dns`` Namespace. The Name of the CR will be set to the wildcard

Reviewer 3: Is an EgressFirewall namespaced? What will the parent object of this be? Or will users create these directly?

> already created. Each CR will be created in the ``openshift-dns`` Namespace. The Name of the CR will be set to the wildcard

Author: Yes. EgressFirewall is namespaced and it can be created by the cluster admin users. https://docs.openshift.com/container-platform/4.12/networking/ovn_kubernetes_network_provider/configuring-egress-firewall-ovn.html

> already created. Each CR will be created in the ``openshift-dns`` Namespace. The Name of the CR will be set to the wildcard

Reviewer 3: Then I would expect the `EgressFirewallDNSName`, if that's what we are calling it, to be owned by the `EgressFirewall` that caused it to be created, which means it should also be namespaced.

DNS name after replacing the ``*`` with ``wildcard`` (barring any trailinng `.`). For example, if the wildcard DNS name is
``*.example.com``, then the Name of the corresponding CR will be ``wildcard.example.com``. This is done to adhere to the

> [Kubernetes object naming validations](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).

Author: The Name field is set to the DNS name so that there's one to one mapping with the ``DNSName`` CR and a DNS name and the CR can be fetched by the CoreDNS plugin just by using the DNS name whenever the lookup happens. I am not sure how this can be achieved without directly using the DNS name. The problem comes with wildcard DNS name.

If we have one ``DNSName`` for each EgressFirewall and the ``.status`` of ``DNSName`` containing all the details of the IPs, would it not be better to just add the information about the IPs in the ``.status`` of the EgressFirewall itself and not create a new CRD?

> [Kubernetes object naming validations](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).

Reviewer 2: > the CR can be fetched by the CoreDNS plugin just by using the DNS name whenever the lookup happens

but the plugin needs to be keeping a cache of all existing DNSName objects anyway, so it could just keep a "name → DNSName object" map in addition to that

> If we have one DNSName for each EgressFirewall and the .status of DNSName containing all the details of the IPs, would it not be better to just add the information about the IPs in the .status of the EgressFirewall itself and not create a new CRD?

yes, maybe...

> [Kubernetes object naming validations](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).

Reviewer 2: (Though that would imply that the plugin was EgressFirewall-specific and couldn't be used for other things.)

> [Kubernetes object naming validations](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).

Reviewer: I have already asked that wildcard dns not be a part of this e.p.  It is not in the RFE.

> [Kubernetes object naming validations](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).

Reviewer 3: I have to agree with @danwinship here in that I think having the DNS name in the CR spec rather than the CR name is a better way to go, it eliminates the complexity of wildcard substitutions since then we can have some much more explicit structure, perhaps something like below
```
spec:
  dns:
    domains:
    - a.example.com
    - b.example.com
    - c.example.com
    sudomains:
    - d.example.com // This would then be treated as *.d.example.com
```
The idea of one DNSName for each EgressFirewall then begs the question, what about one DNSName per cluster? Do we need to separate it at all if all it's doing is storing resolved addresses? Do we ever need more than one?

> [Kubernetes object naming validations](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).

Author: Having a single `DNSName` per cluster will probably have some impact on the performance when looking up DNS names in the status of the object. 

Creating a `DNSName` for each `EgressFirewall` will introduce duplication of information as same DNS name can be used in multiple `EgressFirewalls`. So, the IPs associated with the DNS name have to be updated in all the corresponding `DNSName` objects.

I am more inclined towards having a `DNSName` object for each DNS name (regular and wildcard). The name of the `DNSName` can be a hash value and the spec will contain the corresponding DNS name. Going through the status of this `DNSName` object will have lesser performance impact.

> [Kubernetes object naming validations](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).

Author: For example, if there are 2 EgressFirewall DNS rules, one for `www.redhat.com` and another for `*.openshift.com`, then there will be 2 `DNSName` CRs for each one of them. The  `IsRegular` and `IsWildcard` fields won't be required. Whenever, a lookup happens for `www.redhat.com`, CoreDNS will only fetch the CR corresponding to it and if required update the status. Similarly, if a lookup happens for any subdomain under `*.openshift.com`, then will only fetch the CR corresponding to it and if required update the status.

NOTE: The status of a regular DNS name will only contain one item in the `resolvedNames` field corresponding to the regular DNS name, whereas for a wildcard DNS name the `resolvedNames` will contain one item for the wildcard DNS name and additional items any other subdomains under it, if any.
```yaml
...
metadata:
  name: dns-596764f49c
  namespace: openshift-dns
spec:
  name: "www.redhat.com"
status:
  resolvedNames:
  -  dnsName: "www.redhat.com"
     info:
     - ...
```

```yaml
...
metadata:
  name: dns-747487bbfb
  namespace: openshift-dns
spec:
  name: "*.openshift.com"
status:
  resolvedNames:
  -  dnsName: "*.openshift.com"
     info:
     - ...
  -  dnsName: "docs.openshift.com"
     info:
     - ...
```

> [Kubernetes object naming validations](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).

Author: >Having many objects subscribed will result in far more events hitting the informer than a single shared object, I think in terms of informers a single object would be more efficient.

I think only the create and delete events will be the additional ones. However, the number of update events will be the same as update will happen whenever there's a change in the IPs for a DNS name, whether we use a single object or one object each for each DNS name. The additional impact will be proportional to the number of unique DNS names used in the EgressFirewall rules.

The performance impact that I am thinking about is the time spent to traverse through the status of the CR. The `resolvedNames` field in the status is a slice and if it contains all the DNS names used in EgressFirewall rules, including those which match wildcard DNS names, then every time OVN-K will have to go through the entire slice to get the latest update in the IP addresses. It'll also have spend time to check whether IPs for each DNS name changed or not and update only for the ones that changed. If we have a CR for each DNS name, then OVN-K will immediately know, during an update event, the IPs for which DNS name changed and take appropriate actions. Same for the CoreDNS plugin when it has to update the status of the single object.

Another issue is more from the perspective of parallel updates to the single object by CoreDNS and OVN-K. It was suggested here (https://github.com/openshift/enhancements/pull/1335#discussion_r1102515162) to have a feedback from OVN-K to CoreDNS. If we have a single object, parallel updates may cause some race condition problems with out of order updates. Additionally, CoreDNS pods run as DaemonSets and each pod runs independent of the other. Thus, more than one pod may receive DNS lookup requests for different DNS names simultaneously. When updating the status of the single object chances are there that one update may be done over the other one and some updates may get lost because of this.

> Is CoreDNS fetching the CR directly? Will it be using an informer?

I was planning to have both an informer for the create and delete events in the CoreDNS plugin. But when updating the object, it would be first fetched then the status would be checked with the current IPs and lookup time information. If it requires an update then update the status. OVN-K will be creating and deleting the CRs, so it will be using an informer for the update event to get the changes in the IP information of a DNS name.

> [Kubernetes object naming validations](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).

Reviewer 6: > The performance impact that I am thinking about is the time spent to traverse through the status of the CR. The `resolvedNames` field in the status is a slice and if it contains all the DNS names used in EgressFirewall rules, including those which match wildcard DNS names, then every time OVN-K will have to go through the entire slice to get the latest update in the IP addresses. It'll also have spend time to check whether IPs for each DNS name changed or not and update only for the ones that changed. If we have a CR for each DNS name, then OVN-K will immediately know, during an update event, the IPs for which DNS name changed and take appropriate actions. Same for the CoreDNS plugin when it has to update the status of the single object.

This makes sense to me.  Keep in mind that there is potential fanout:  An EgressFirewall can have rules with many DNS names or wildcards, and a wildcard can result in an EgressFirewallDNSName with many values under `status.resolvedNames`.  

> Another issue is more from the perspective of parallel updates to the single object by CoreDNS and OVN-K. It was suggested here ([#1335 (comment)](https://github.com/openshift/enhancements/pull/1335#discussion_r1102515162)) to have a feedback from OVN-K to CoreDNS. If we have a single object, parallel updates may cause some race condition problems with out of order updates. Additionally, CoreDNS pods run as DaemonSets and each pod runs independent of the other. Thus, more than one pod may receive DNS lookup requests for different DNS names simultaneously. When updating the status of the single object chances are there that one update may be done over the other one and some updates may get lost because of this.

Agreed.  We will have many CoreDNS pods updating the EgressFirewallDNSName objects to update `status.resolvedNames` (by default, a cluster has 1 DNS pod per node), and we don't want these pods to be constantly in conflict with each other, which would cause frequent retries and unnecessarily overhead for both the CoreDNS plugin and kube-apiserver.

* The ``egressfirewall`` CoreDNS plugin will intercept the request and the response for the DNS lookups from all the pods. If
the regular DNS name in the lookup matches with a wildcard DNS name, then the ``egressfirewall`` plugin will update the
``.status`` of the corresponding `DNSName` CR with the regular DNS name and the corresponding current IPs along with the TTL
and the next lookup time based on the TTL. However, this update will only take place if there is a change in the exisiting IP
addresses or the next lookup time or both for the DNS name.
* The OVN-K master will watch the ``DNSName`` CRs. When the ``.status`` of a ``DNSName`` CR is updated, the OVN-K master
will update the ``AddressSet`` for the wildcard DNS name, which is linked with the ACL rule for the corresponding EgressFirewall(s).
* The OVN-K master will also store the regular DNS name and the corresponding current IPs along with the TTL and the next time
to lookup. Based on the next time to lookup, the OVN-K master will follow the same method as that of the regular DNS names to get the
latest IPs and TTL information.

#### Deletion

##### Regular DNS name

> * An OpenShift cluster administrator deletes an EgressFirewall resource for a Namespace containing rul(e) for regular DNS name(s).

Author: Yes. That was a typo. Already updated.

rules. The OVN-K master will also delete the corresponding ``DNSName`` CRs, only if the ``.spec.isregular`` field is set to true and
the ``.spec.iswildcard`` field is set to false. If both the fields are set to true, then the CR will not be deleted and the
``.spec.isregular`` field will be set to false.

##### Wildcard DNS name
* An OpenShift cluster administrator deletes an EgressFirewall resource for a Namespace containing rul(e) for wildcard DNS name(s).
* The OVN-K master will delete the ACL rule corresponding to the EgressFirewall.
* The OVN-K master will then check if the same wildcard DNS names are also used in the EgressFirewall rules in other Namespaces. If
they are not used, then the OVN-K master will delete the corrresponding ``AdressSets`` for each of the DNS names in the EgressFirewall
rules. The OVN-K master will also delete the corresponding ``DNSName`` CRs, only if the ``.spec.isregular`` field is set to false and
the ``.spec.iswildcard`` field is set to true. If both the fields are set to true, then the CR will not be deleted and the
``.spec.iswildcard`` field will be set to false. From the ``.status`` field all the other DNS names' details will be removed and

> only the details of the DNS name will be kept which matches the Name field of the CR.

Reviewer 2: This is just a mess, and AFAICT it only matters in the case where you have both `*.example.com` and `wildcard.example.com`? It would be much better to redesign the API to not use the same object for both cases.

---
title: coredns-egressfirewall-integration
authors:
  - '@arkadeepsen'
reviewers:
  - '@Miciah'
  - '@danwinship'
  - '@JoelSpeed'
  - '@TrilokGeer'
approvers:
  - '@Miciah'
  - '@danwinship'
api-approvers:
  - '@JoelSpeed'
creation-date: 2023-01-31
last-updated: 2023-01-31
tracking-link:
  - https://issues.redhat.com/browse/CFE-748
see-also:
  - N.A.
replaces:
  - N.A.
superseded-by:
  - N.A.
---

# Improve CoreDNS Integration with EgressFirewall

## Summary

This enhancement improves the integration of CoreDNS with EgressFirewall. With this improved
integration, EgressFirewall will be able to better support DNS names whose IPs change dynamically
and also will be able to provide support for wildcard DNS names.

## Motivation

> Currently, EgressFirewall (OVN-K master) does a DNS lookup of a DNS name based on a default TTL or the

Author: Yes. It is documented here: https://docs.openshift.com/container-platform/4.12/networking/ovn_kubernetes_network_provider/configuring-egress-firewall-ovn.html#domain-name-server-resolution_configuring-egress-firewall-ovn

TTL of the previous lookup. OVN-K master then updates the underlying ``AddressSet`` for the DNS name
referenced by the corresponding ACL rule of the EgressFirewall rule containing the DNS name. However,

> if a pod, belonging to the same Namespace as that of the EgressFirewall, does a DNS lookup and is able
> to get the response before the OVN-K master then the pod will be incorrectly allowed/denied access to
> the DNS name. The integration between CoreDNS and EgressFirewall needs to be improved to avoid such a

Reviewer: ```suggestion
if a pod, belonging to the same Namespace as that of the EgressFirewall, does a DNS lookup and gets a different response than the OVN-K master has, then the pod will be incorrectly allowed/denied access to
the host. The integration between CoreDNS and EgressFirewall needs to be improved to avoid such a
```

scenario.

If an administrator wants to specifically allow/deny access to all subdomains then currently the
administrator has to add all subdomains in the EgressFirewall rules. This becomes difficult when
subdomains are dynamically added/removed as each one has to be added/removed individually from the
EgressFirewall rules. Currently, wildcard DNS names are not supported in EgressFirewall. However,
even if the support is added to EgressFirewall, the integration between CoreDNS and EgressFirewall

> needs to be improved as wildcard DNS names cannot be directly looked up through a DNS query.

Reviewer: On the contrary, there is actually another RFE for wildcard dns here: https://issues.redhat.com/browse/RFE-1384  Even if you ignore the other RFE, there are numerous issues with wildcard DNS names in an egress firewall, and you haven't addressed them in this EP.  Because of this, I cannot approve the proposal as is.

> needs to be improved as wildcard DNS names cannot be directly looked up through a DNS query.

Author: > there are numerous issues with wildcard DNS names in an egress firewall, and you haven't addressed them in this EP

Can you please mention the issues which are not addressed in this EP?

Regarding the separate RFE, @mcurry-rh and @Deepthidharwar would be in a better position to respond to this. cc @danwinship

> needs to be improved as wildcard DNS names cannot be directly looked up through a DNS query.

Reviewer 2: PM wanted to implement the "wildcard DNS in egress firewall" feature, and the only way to do that is to change the way that all egress firewall DNS works. Since there was already an older RFE for "change the way that all egress firewall DNS works", it seemed to make more sense to call this part of that rather than vice versa. Maybe we should have done it the other way around.

Anyway, regardless of what is linked where, the actual requested-by-PM feature here is "wildcard DNS in egress firewall", so the enhancement needs to include that functionality (in some way that doesn't cause problems, obviously). If we can't figure out any way to implement wildcard DNS in egress firewall in a good way, then the enhancement would be dropped, because there's not enough customer need for _just_ "CoreDNS/EgressFirewall integration without wildcards".

> needs to be improved as wildcard DNS names cannot be directly looked up through a DNS query.

Reviewer: Security issues of using DNS wildcard should be part of the evaluation of accepting the RFE, but I'll do my best:
1.  a bad actor could just keep generating new dns names when there is a wildcard option to allow any new subdomains in the rules.  In other words, for a wildcard dns you would need to list all the deny lines first, then list an allow line for the wildcard subdomain.  You can deny a.b.com and allow *.b.com, but a.b.com can just change its name or publish an alias of aaa.b.com to get around it, popping a hole in the firewall
2. the pod could go directly to an upstream nameserver for resolution, bypassing the synchronization effort of the new feature
3. the pod/application could have prior knowledge of an IP address and use that instead of a dns name

> needs to be improved as wildcard DNS names cannot be directly looked up through a DNS query.

Author: > * a bad actor could just keep generating new dns names when there is a wildcard option to allow any new subdomains in the rules.  In other words, for a wildcard dns you would need to list all the deny lines first, then list an allow line for the wildcard subdomain.  You can deny a.b.com and allow *.b.com, but a.b.com can just change its name or publish an alias of aaa.b.com to get around it, popping a hole in the firewall

The user creating the EgressFirewall object should be aware of the matching order: https://docs.openshift.com/container-platform/4.12/networking/ovn_kubernetes_network_provider/configuring-egress-firewall-ovn.html#domain-name-server-resolution_configuring-egress-firewall-ovn

Additionally, as discussed previously from the point of view of security issues only DNS allow rules should be supported. There should first be a deny all rule then DNS allow rules: https://github.com/openshift/enhancements/pull/1335#discussion_r1101797917

> * the pod could go directly to an upstream nameserver for resolution, bypassing the synchronization effort of the new feature

DNS operator sets up the pods to use CoreDNS for resolving DNS names: https://docs.openshift.com/container-platform/4.12/networking/dns-operator.html#nw-dns-operator_dns-operator

Even if this is possible, this problem would be solved by the approach mentioned above (deny all rule followed by DNS allow rules). Thus the request to the upstream nameserver will not go through.

> * the pod/application could have prior knowledge of an IP address and use that instead of a dns name

Again, this problem can be solved by the same approach (deny all rule followed by DNS allow rules). The traffic to the IP will not go through.

> needs to be improved as wildcard DNS names cannot be directly looked up through a DNS query.

Reviewer: From the doc: 

> The egress firewall policy rules are evaluated in the order that they are defined, from first to last. The first rule that matches an egress connection from a pod applies. Any subsequent rules are ignored for that connection.

This means if you had a deny wildcard rule first, it would not pass to any of the subsequent rules.

> needs to be improved as wildcard DNS names cannot be directly looked up through a DNS query.

Author: It should be a IP deny rule on `0.0.0.0/0` and not a deny wildcard DNS rule. DNS rules will not be used for deny rules as mentioned above. The example CR explains how the deny rule needs to be added: https://docs.openshift.com/container-platform/4.12/networking/ovn_kubernetes_network_provider/configuring-egress-firewall-ovn.html#egressnetworkpolicy-example_configuring-egress-firewall-ovn

It is also explained in the `Important` block in this section: https://docs.openshift.com/container-platform/4.12/networking/ovn_kubernetes_network_provider/configuring-egress-firewall-ovn.html#nw-egressnetworkpolicy-about_configuring-egress-firewall-ovn

### User Stories

* As an OpenShift cluster administrator, I want to add regular DNS names to EgressFirewall rules, so that I can allow/deny
access to them even if the IPs associated with the corresponding DNS records change dynamically.
* As an OpenShift cluster administrator, I want to add wildcard DNS names to EgressFirewall rules, so that I can
allow/deny access to all the subdomains belonging to the wildcard DNS names.
* As an OpenShift engineer, I want to add a new Custom Resource, so that IPs and TTLs of the DNS names which are used
in EgressFirewall rules can be tracked.
* As an OpenShift engineer, I want to add a new plugin to CoreDNS, so that DNS lookups of DNS names used in EgressFirewall
rules can be inspected and the current IPs and TTLs can be tracked in the corresponding new CR.
* As an OpenShift engineer, I want to modify Cluster DNS operator, so that CoreDNS can be deployed with the
new plugin enabled.

### Goals

* Support update of the ``AddressSet`` referenced by the ACL rule corresponding to an EgressFirewall rule containing

> a DNS name if the IPs associated with it change dynamically.

Author: I will remove the word "dynamically". It wanted to mean that the IPs associated to the DNS name are not static. But I think it creates confusion.

* Create CRs for each unique DNS name used in the EgressFirewall rules and use the CRs to track the current
IPs and the corresponding TTL information.

### Non-Goals

* Support additional DNS resolution functionality in the new CoreDNS plugin.
* Support allowing/denying of DNS lookups (core EgressFirewall functionality) in the new CoreDNS plugin.

## Proposal

This enhancement proposes to introduce a new CoreDNS plugin (``egressfirewall``) and a new Custom Resource
(``DNSName``) to improve the integration of CoreDNS with EgressFirewall. The ``DNSName`` CR will be created
for each unique DNS name (both regular and wildcard DNS names) used in the EgressFirewall rules. This CR will
be used to store the DNS name along with the current IPs and the correspodning TTL and the next lookup time
based on the TTL. The OVN-K master will be responsible for the creation of the ``DNSName`` CRs. The ``DNSName``
CR is meant for communication between CoreDNS and OVN-K master. Thus it cannot be created or modified by an
OpenShift cluster adminisrator.

The new plugin will inspect each DNS lookup and the corresponding response for the DNS lookup from other

> plugins. If the DNS name in the query matches any ``DNSName`` CR(s) (regular or wildcard or both), then the

Reviewer: The new plugin should only be used by EgressFirewall, not any other lookup mechanisms.

Author: The DNS query will always be processed by the new plugin as per the processing logic of CoreDNS plugin chain (https://coredns.io/2017/06/08/how-queries-are-processed-in-coredns/). Thus, the matching of the DNS name in the query needs to be done with the `DNSName` CRs and if it matches only then the actual logic of the new plugin will be executed.

Author: > * What is the impact if there are no DNSName CRs?  Is it only the time it takes to execute `if len(dn.Items) == 0 { /* Execute the next plugin. */}`?

That's correct.

> * What is the impact if there are DNSName CRs but none that matches the query?  Is it the cost of a map lookup?  Is the cost greater when rules use wildcards?  I'm curious what data structure the plugin will use for looking up the DNSName CR for a query.

We can have 2 separate maps for storing regular and wildcard DNS names. The check should happen for both types of DNS names. For wildcard DNS name, the check should happen for each subdomain as well. For example, if a DNS lookup is done for `sub1.sub2.example.com`, then the lookup needs to happen for `*.sub2.example.com` and `*.example.com`. So for wildcards the cost will be higher.

> * What is the impact if some DNSName CR matches the query?  Presumably the status update can happen asynchronously, but does the plugin force a delay while OVN-K updates?

In the current version of the EP, there is no delay that is enforced while waiting for OVN-K updates. However, it has been suggested in this https://github.com/openshift/enhancements/pull/1335#discussion_r1102515162 to wait for OVN-K updates before responding to DNS queries.

Reviewer 6: > For wildcard DNS name, the check should happen for each subdomain as well. For example, if a DNS lookup is done for `sub1.sub2.example.com`, then the lookup needs to happen for `*.sub2.example.com` and `*.example.com`. So for wildcards the cost will be higher.

In https://github.com/openshift/enhancements/pull/1335#discussion_r1157154525, you said that a `*` would match only one label, so I would expect that for a query for `sub1.sub2.example.com`, the plugin would look up `sub1.sub2.example.com` in the regular map and `*.sub2.example.com` (or simply `sub2.example.com`) in the wildcard map.  Is that what you meant, or am I misunderstanding something?

Author: Sorry. My bad. It should only try to match `*.sub2.example.com` for wildcard as `*` will only match one label. Thanks for pointing that out.

Reviewer: @arkadeepsen following up on this discussion, is there any conclusion/update to Miciah's question on :

> What is the impact if some DNSName CR matches the query? Presumably the status update can happen asynchronously, but does the plugin force a delay while OVN-K updates?

It seems like this might need update.  If possible, just answer the question here instead of reference to another part of the doc.

Author: In a meeting with @danwinship, @Miciah and @npinaeva, it was decided that the CoreDNS plugin will not wait for the OVN-K updates. So there won't be any forced delay on the plugin from the OVN-K side.

plugin will update the ``.status`` of the matching ``DNSName`` CR(s) with the DNS name along with the IPs and
the corresponding TTL and the next lookup time based on the TTL. The OVN-K master will watch the ``DNSName``
CRs. Whenever the IPs are updated for a ``DNSName`` CR, the OVN-K master will update the the underlying ``AdressSet``
referenced by the ACL rule(s) for the corresponding EgressFirewall rule(s).

OVN-K master will keep track of the TTL (or next lookup time) for each regular DNS name and send a DNS lookup

> query to CoreDNS when the minimum TTL expires. However, for a wildcard DNS name a DNS lookup cannot be performed

Author: Yes. This the current behavior of OVN-K to trigger the lookups based on the TTL values. I have added it here as this behavior is still required for supporting DNS rules in EgressFirewall (both regular and wildcard).

> query to CoreDNS when the minimum TTL expires. However, for a wildcard DNS name a DNS lookup cannot be performed

Author: Currently in OVN-K both TTL and next lookup time is stored in data structure used for DNS names. If TTL is not used anywhere apart from calculating next lookup time then it can be removed.

directly on the DNS name as it will not return any IP. Thus, the lookups will be performed on the DNS names
which are updated in the ``.status`` of the corresponding wildcard ``DNSName`` CRs.

> ### Workflow Description

Reviewer 2: I think this section would be easier to understand if the API description came first.

The workflows for Create, Delete and Update events for EgressFirewall related to DNS names are explained in this section.

#### Create/Update

##### Regular DNS name

* An OpenShift cluster administrator creates/updates an EgressFirewall resource for a Namespace and adds rule(s) containing regular
DNS name(s).
* The OVN-K master will create corrresponding `DNSName` CRs for each of the DNS names in the EgressFirewall rules, if not
already created. Each CR will be created in the ``openshift-dns`` Namespace and the Name of the CR will be same as the DNS
name (barring any trailinng `.`). The ``.spec.isregular`` field of the CR will be set to true, even if it already exists.
* The OVN-K master will then perform DNS lookup for each of the regular DNS names added to the EgressFirewall rules.
* The ``egressfirewall`` CoreDNS plugin will intercept the request and the response for the DNS lookup for each of the
regular DNS names.

> * As these DNS names have corresponding ``DNSName`` CRs, the ``egressfirewall`` plugin will update the ``.status`` of

Reviewer: The CoreDNS plugin shouldn't be updating the DNSName CRs, the egress firewall should be.  Afaict, the DNSName CRs are relevant only to the egress firewall.

on the TTL. However, this update will only take place if there is a change in the exisiting IP
addresses or next time to lookup or both for the DNS name.
* The OVN-K master will watch the ``DNSName`` CRs. When the ``.status`` of a ``DNSName`` CR is updated, the OVN-K master
will update the ``AddressSet`` for the DNS name, which is linked with the ACL rule(s) for the corresponding EgressFirewall
rule(s).
* The OVN-K master will also receive the response of the DNS lookup query for the DNS name. The OVN-K master will check
the corresponding ``DNSName`` CR's ``.status`` and if the next lookup time in the status is greater than the next lookup
time based on the received TTL, then the corresponding CR's ``.status`` will be updated. The corresponding ``AddressSet``
will also be updated.

##### Wildcard DNS name

* An OpenShift cluster administrator creates/updates an EgressFirewall resource for a Namespace and adds rule(s) containing wildcard
DNS name(s).
* The OVN-K master will create corrresponding `DNSName` CRs for each of the wildcard DNS names in the EgressFirewall rules, if

> not already created. Each CR will be created in the ``openshift-dns`` Namespace. The Name of the CR will be set to the wildcard

Author: No specific reason. Will change to small letters.

DNS name after replacing the ``*`` with ``wildcard`` (barring any trailinng `.`). For example, if the wildcard DNS name is
``*.example.com``, then the Name of the corresponding CR will be ``wildcard.example.com``. This is done to adhere to the
[Kubernetes object naming validations](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).
The ``.spec.iswildcard`` field of the CR will be set to true, even if it already exists.
* The OVN-K master will not perform any DNS lookup for the wildcard DNS names directly.

> * The ``egressfirewall`` CoreDNS plugin will intercept the request and the response for the DNS lookups from all the pods. If

Reviewer: Do you mean that this plugin will now intercept all the pods' dns requests?  How does a dns request get directed to the plugin?

Author: CoreDNS uses plugin chain to process a DNS query. The plugin chain is defined in the `plugin.cfg` file and which plugins are enabled is defined in the CoreFile. https://coredns.io/2017/06/08/how-queries-are-processed-in-coredns/

Reviewer: I'm talking about your new plugin.

Author: The process of adding it to the plugin chain is describe here: https://github.com/openshift/enhancements/pull/1335/files#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R254-R261

> the regular DNS name in the lookup matches with a wildcard DNS name, then the ``egressfirewall`` plugin will update the

Author: The `*` will match only one label. Also only a single `*` can be used at the beginning of the wildcard DNS name.

``.status`` of the corresponding `DNSName` CR with the regular DNS name and the corresponding current IPs along with the TTL
and the next lookup time based on the TTL. However, this update will only take place if there is a change in the exisiting IP
addresses or the next lookup time or both for the DNS name.
* The OVN-K master will watch the ``DNSName`` CRs. When the ``.status`` of a ``DNSName`` CR is updated, the OVN-K master
will update the ``AddressSet`` for the wildcard DNS name, which is linked with the ACL rule for the corresponding EgressFirewall

> rule(s).

Author: Yes. This has been suggested in this https://github.com/openshift/enhancements/pull/1335#discussion_r1102515162

* An OpenShift cluster administrator deletes an EgressFirewall resource for a Namespace containing rule(s) for regular DNS
name(s) OR updates an EgressFirewall resource for a Namespace and deletes rule(s) containing regular DNS name(s).
* The OVN-K master will delete the ACL rule(s) corresponding to the EgressFirewall rule(s) containing the regular DNS name(s).
* The OVN-K master will then check if the same regular DNS names are also used in the EgressFirewall rules in other Namespaces. If
they are not used, then the OVN-K master will delete the corrresponding ``AdressSet`` for each of the DNS names in the EgressFirewall
rules. The OVN-K master will also delete the corresponding ``DNSName`` CRs, only if the ``.spec.isregular`` field is set to true and
the ``.spec.iswildcard`` field is set to false. If both the fields are set to true, then the CR will not be deleted and the
``.spec.isregular`` field will be set to false.

##### Wildcard DNS name

* An OpenShift cluster administrator deletes an EgressFirewall resource for a Namespace containing rule(s) for wildcard DNS name(s)
OR updates an EgressFirewall resource for a Namespace and deletes rule(s) containing wildcard DNS name(s).

> * The OVN-K master will delete the ACL rule(s) corresponding to the EgressFirewall rule(s) containing the wildcard DNS name(s).

Reviewer 3: Are these separated somehow so that this logic will exclude subdomains that are covered both explicitly and within a wildcard?

* The OVN-K master will then check if the same wildcard DNS names are also used in the EgressFirewall rules in other Namespaces. If
they are not used, then the OVN-K master will delete the corrresponding ``AdressSet`` for each of the DNS names in the EgressFirewall
rules. The OVN-K master will also delete the corresponding ``DNSName`` CRs, only if the ``.spec.isregular`` field is set to false and
the ``.spec.iswildcard`` field is set to true. If both the fields are set to true, then the CR will not be deleted and the
``.spec.iswildcard`` field will be set to false. From the ``.status`` field all the other DNS names' details will be removed and
only the details of the DNS name will be kept which matches with the Name field of the CR.

> #### Variation [optional]

Reviewer 2: if you're not using this section you can delete it

Author: The EP template mentioned not to delete any section even if it's not used.

### API Extensions

The validation of ``DNSName`` field in ``EgressFirewallDestination`` will be updated to accept wildcard DNS names as well.
It will be updated from ``^([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$`` which accepts only regular DNS names to
``^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$``.

````go
// EgressFirewallDestination is the endpoint that traffic is either allowed or denied to
type EgressFirewallDestination struct {
	// ..

> 	// dnsName is the domain name to allow/deny traffic to. If this is set, cidrSelector must be unset.

Author: Ok. I will add the details. I have clarified it in this https://github.com/openshift/enhancements/pull/1335#discussion_r1157154525

	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$

> 	DNSName string `json:"dnsName,omitempty"`

Reviewer 3: What is the structure of this struct already? Having fields that are mutually exclusive makes me think this would benefit from being a discriminated union. We will at a minimum want to validate that the two fields are not set at the same time. This will require a CEL validation rule and an integration test within the API repo

Author: This [field](https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/crd/egressfirewall/v1/types.go#L68-L81) is part of EgressFirewall and has the same name. There is a suggestion to change the CR name from `DNSName` to `EgressFirewallDNSName` which will remove the confusion.

Author: [openshift/ovn-kubernetes](https://github.com/openshift/ovn-kubernetes/tree/master) is downstream fork of  [ovn-org/ovn-kubernetes](https://github.com/ovn-org/ovn-kubernetes/tree/master). The `EgressFirewallDestination` struct is copied [here](https://github.com/openshift/ovn-kubernetes/blob/master/go-controller/pkg/crd/egressfirewall/v1/types.go#L68-L81) in openshift/ovn-kubernetes repo.

}
````

> The following ``DNSName`` CRD will be added to the ``k8s.ovn.org`` api-group.

Reviewer 2: While _we_ are using the CoreDNS plugin with ovn-kubernetes, it's not logically bound to ovn-kubernetes. I think this should be something like `dns.openshift.io`

> The following ``DNSName`` CRD will be added to the ``k8s.ovn.org`` api-group.

Reviewer 6: If the group name makes it clear that this is for OVN or for EgressFirewall, then the CRD name doesn't necessarily need to indicate it.  So, for example, if the group name is "k8s.ovn.org" or "egressfirewall.operators.openshift.io" or whatever, then the qualified resource name ends up as "dnsnames.k8s.ovn.org" or "dnsnames.egressfirewall.operators.openshift.io", which makes it clear that the resource is related to the OVN-CoreDNS integration for EgressFirewall.

> The following ``DNSName`` CRD will be added to the ``k8s.ovn.org`` api-group.

Author: I have updated the EP with the following changes that were discussed:
- The CoreDNS plugin will not wait for OVN-K masters to update the underlying ACL rules
- OVN-K cluster manager will create and delete the new CR.
- A new controller will watch the CR and perform the re-resolution of the DNS names added to the `.status` of the CRs.
- The api definition of the new CRD is updated.

@Miciah @danwinship @npinaeva @JoelSpeed PTAL and let me know if any other changes are required.

Some of the points which are still kind of unresolved:
- What should be the name of the new CRD?
As of now I have used `EgressFirewallDNSName`.
- Which apigroup should it be added to?
As of now I have used `dns.openshift.io` as per the suggestion here: https://github.com/openshift/enhancements/pull/1335#discussion_r1102096437
- Should the new CRD be cluster-scoped or namespace-scoped?
As of now it is a namespace-scoped CR. There was a previous suggestion to keep it namespace-scoped itself: https://github.com/openshift/enhancements/pull/1335#discussion_r1101825551
- What should be the name of the new CoreDNS plugin?
As of now I have used `egressfirewall`. However, if we are changing the CRD name and apigroup then we probably have to change the name of the plugin as well.

> The following ``DNSName`` CRD will be added to the ``k8s.ovn.org`` api-group.

Author: As per today's discussions, the name of the CRD is decided to be `DNSNameResolver` and it'll belong to `network.openshift.io` apigroup. The name of the new plugin will be `ocp_dnsnameresolver`.

````go
// DNSName describes a DNS name used in a EgressFirewall rule.
type DNSName struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// Specification of the desired behavior of DNSName.
	Spec DNSNameSpec `json:"spec"`
	// Observed status of DNSName
	// +optional
	Status DNSNameStatus `json:"status,omitempty"`
}

// DNSNameSpec is a desired state description of DNSName.
type DNSNameSpec struct {
	// IsRegular denotes whether the resource is representing a regular DNS name.
	// Cannot be false if IsWildcard is also false.
	// +optional

> 	IsRegular bool `json:"isregular,omitempty"`

Reviewer: I think there is another solution, don't do wildcard dns.  It is not a part of the RFE and not a good idea for reasons already mentioned.

	// IsWildcard denotes whether the resource is representing a wildcar DNS name.
	// Cannot be false if IsRegular is also false.
	// +optional

> 	IsWildcard bool `json:"iswildcard,omitempty"`

Reviewer 3: Booleans don't generally make sense in Kubernetes APIs, especially since they tend to evolve over time to become combination bools that have 3 or 4 valid combinations as well as other invalid comibinations.

Here you have a perfect example of a paired boolean, "Cannot be false if IsRegular is also false.", defining behaviour of fields based on the behaviour of other fields is typically confusing.

If you want to have a notion of regular vs wildcard, you would be better using an enum that allows users to specify which type of dns name this is
```
type: Regular | Wildcard
```

}

type DNSNameStatus struct {
	// The list of matching DNS names and their corresponding IPs along with TTL and next
	// time to DNS lookup.

> 	Items []DNSNameStatusItem `json:"items"`

Reviewer 3: Why items? 
Maybe better
```
resolvedNames:
- name: bla
  info:
  - ...
```

Author: It has been suggested in this https://github.com/openshift/enhancements/pull/1335#discussion_r1156644387 to change it to `records`.

}

type DNSNameStatusItem struct {
	DNSName string        `json:"dnsname"`

> 	Info    []DNSNameInfo `json:"info"`

Reviewer 3: Why does this need to be a slice? Are the TTL and lookup times related to the name or the IP addresses? I would have expected to look up the name and receive a list of IP addresses, so by my intuition, the TTL is based on the name not the IPs as it is structured here

Author: The response of a DNS lookup contains TTL for each IP associated with the DNS name. The `Msg` format can be found [here](https://github.com/openshift/coredns/blob/master/vendor/github.com/miekg/dns/msg.go#L108-L116). The `Answer` field is a slice. Assuming the answer is an `A` type record, then the format can be found [here](https://github.com/openshift/coredns/blob/master/vendor/github.com/miekg/dns/types.go#L713-L717) and the `RR_Header` format can be found [here](https://github.com/openshift/coredns/blob/master/vendor/github.com/miekg/dns/dns.go#L66-L73).

Reviewer 3: Can you make sure that the answers to my questions (which I think you've answered), end up within the godocs for these fields and structs if they are staying like this within the API

}

type DNSNameInfo struct {

> 	IP             string `json:"ip"`

Reviewer: Instead of a string, use a valid IP address representation like net.IP or IPAddr.

Author: Ok. Will update the type.

> 	TTL            string `json:"ttl"`

Author: Ok. Wil change the type to `int64`.

> 	NextLookupTime string `json:"nextlookuptime"`

Reviewer 3: Typically we include a `lastUpdated` rather than `nextUpdate`, you can calculate the next lookup by using the TTL, but it's also useful to see when was it last updated, for example, are you expecting the update to happen only after the data expires? I would probably expect the logic to see if we are close to the TTL based on when it was last updated and update it _before_ it expires right? The current API makes me think it's being updated after it expires

Author: Makes sense. I will use `lastUpdated` instead of `NextLookupTime`.

}
````

### Implementation Details/Notes/Constraints [optional]

The implementation changes needed for the proposed enhancement are documented in this section for each of the components.

#### Cluster DNS Operator

Cluster DNS Operator will deploy CoreDNS with the ``egressfirewall`` plugin enabled by adding it to the corefile. As the
plugin will watch and update the ``DNSName`` CRs in the ``k8s.ovn.org`` api-group, proper RBAC permissions will be needed
to be added to the ``ClusterRole`` for CoreDNS.

#### CoreDNS

The new plugin ``egressfirewall`` will be added to CoreDNS. As the plugin will inspect the DNS lookup queries and response from

> other plugins, it needs to be added before the other plugins (namely ``forward`` plugin) in the ``plugin.cfg`` file.

Author: > Will the new plugin go before or after the cache plugin in `plugin.cfg`?

Would it be better to not add the new plugin before cache plugin? As cache plugin won't provide new IPs, we can reduce the time spent in inspecting responses from cache plugin.

> Will it be added to all server blocks, or just the main `.:5353` one in the Corefile?

Should it be added to all server blocks? I was thinking of adding it to only the main `.:5353` server block.

> other plugins, it needs to be added before the other plugins (namely ``forward`` plugin) in the ``plugin.cfg`` file.

Author: > > Would it be better to not add the new plugin before cache plugin? As cache plugin won't provide new IPs, we can reduce the time spent in inspecting responses from cache plugin.
> 
> I'm not sure. If the egressfirewall plugin comes before the cache plugin, then the egressfirewall plugin's logic will run for cached responses. A cached response will be the same as the original response, except that its TTL is reduced by the time it has already been cached. So if I understand correctly, the egressfirewall plugin would update the DNSName CR with the new TTL and next-lookup-time values. That's the only functional difference I can think of between putting egressfirewall first and putting cache first.

As only the TTL values are reduced by the time it has been cached, will the next lookup time change? Also, the status of the DNSName CRs are updated only when the associated IPs change, otherwise it's not updated. Wouldn't it be better to avoid the inspection of the response from the cache plugin by the new plugin as it will not trigger any updates?

> > > Will it be added to all server blocks, or just the main `.:5353` one in the Corefile?
> > 
> > 
> > Should it be added to all server blocks? I was thinking of adding it to only the main `.:5353` server block.
> 
> I think you want it for all server blocks because you want EgressFirewall rules to apply consistently even for DNS names that are resolved by custom upstreams—or is there some reason this wouldn't be expected and desirable?

Ok, it makes sense. Will add it to all the server blocks.

> other plugins, it needs to be added before the other plugins (namely ``forward`` plugin) in the ``plugin.cfg`` file.

Author: There is a potential issue with the `ocp_dnsnameresolver` plugin being placed below `cache` plugin. If a DNS name is looked up and then before the TTL of the corresponding IP addresses expire an EgressFirewall with a DNS rule (or a DNSNameResolver) is created for the same DNS name, then the subsequent DNS lookups, until the TTL expires, will be served by the `cache` plugin. As a result, for the time till the TTL expires, the DNSNameResolver object for the corresponding DNS name will not be updated with the IP addresses. This problem will be resolved if we put the `ocp_dnsnameresolver` plugin before the `cache` plugin.

The ``egressfirewall`` plugin will watch the ``DNSName`` CRs and whenever there is a DNS lookup which matches one of the ``DNSName``

> CRs (either regular or wildcard DNS names or both), then it will update the ``.status`` of the ``DNSName`` CR(s) if there's any change

Author: > what kind of informer, index, or data structure will the plugin use?

I am planning to use SharedIndexInformer. As for data structure, I am planning to use maps to store DNSName CR information. Do you have any suggestions?

> How is the matching done, especially with wildcards?

I have mentioned about this in this https://github.com/openshift/enhancements/pull/1335#discussion_r1157138722. I'll add the details in the EP.

> How do you mitigate performance impact when there are no matching CRs?

I have mentioned about this as well in this https://github.com/openshift/enhancements/pull/1335#discussion_r1157138722. I'll also add these details in the EP.

> Is the update done asynchronously? What happens if there is a conflict during an update?

Once the update to the DNSName CR is completed, then the response to the DNS lookup will be sent. If the update to the DNSName CR fails then the DNS lookup should also fail, otherwise this might cause discrepancy in the information between the pods and OVN-K. We can add retries to the DNSName CR update process.

> CRs (either regular or wildcard DNS names or both), then it will update the ``.status`` of the ``DNSName`` CR(s) if there's any change

Reviewer 6: > As for data structure, I am planning to use maps to store DNSName CR information. Do you have any suggestions?

I was trying to think of a suitable data structure that could accommodate both regular and wildcard names/lookups, but using two maps as you suggested in https://github.com/openshift/enhancements/pull/1335#discussion_r1157138722 is probably better.

will be same as that of the DNS name (for wildcard DNS name the ``*`` will be replaced by ``wildcard``) without any trailing ``.``.
This will follow the [Kubernetes object naming validations](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names).

The OVN-K master will also watch the ``DNSName`` CRs. Whenever the ``.status`` of the CRs will be updated with new IPs and corresponding
TTL information for a DNS name, OVN-K master will update the ``AdressSet`` mapped to the DNS name. This ``AdressSet`` will be linked
to the ACL rule(s) for the EgressFirewall rule(s) in which the DNS name is used. This will ensure that the latest IPs are always updated in
the ``AddressSets``.

The OVN-K master will not directly query for the wildcard DNS names, rather it will query for the DNS names that get added to the
``.status`` of the corresponding ``DNSName`` CR. However, the list of the DNS names to lookup for a wildcard DNS name should also
not become stale if a DNS name belonging its subdomain is removed. To achieve this a retry counter will be used for the DNS name

> lookups. If the lookup fails for a DNS name listed in the ``.status`` of a ``DNSName`` CR for threshold number of times (say 5),

Reviewer 3: I assume this failure is only after the TTL has expired?

> then the DNS name will be removed from the ``.status``.

Author: > Is there only 1 active OVN-K master at a time?

Yes. A leader is elected amongst the masters: https://github.com/openshift/ovn-kubernetes/blob/master/go-controller/cmd/ovnkube/ovnkube.go#L387-L394

> Is OVN-K using the cluster DNS service?

Are you asking about  if cluster DNS service is used for resolving DNS names used in Egress Firewall? 

> Is there any risk of weird feedback loops?

When OVN-K observes that there's an update to the DNSName CR's status, it'll just use the new IPs and update the underlying AddressSet for the EgressFirewall rules. It'll not trigger a DNS lookup from the side of OVN-K master. OVN-K will only start a DNS lookup when the TTL expires. So, there won't be a risk of such feedback loop.

### Risks and Mitigations

### Drawbacks

* Whenever there's a change in the IPs or the next lookup time for a DNS name, the additional step of updating the

> related ``DNSName`` CRs will be executed. This might add some delay to the DNS lookup process. However, this will only

Author: Currently this delay only refers to the update of the DNSName CR and not the OVN-K updates to propagate.

happen whenever there's a change in the DNS information.

## Design Details

### Open Questions [optional]

### Test Plan

* This enhancement will be tested through e2e tests by adding EgressFirewall rules containing regular DNS names

> and wildcard DNS names.

Reviewer 6: Using openshift/origin makes sense to me.

* Testing the feature where IPs are dynamically changed may be a little bit tricky as this will probably include
creation of DNS records and then changing the IP adresses of the DNS records through the e2e tests.

### Graduation Criteria

> This is a user facing change and will directly go to GA. This feature requires an update to Openshift Docs.

Reviewer 2: I don't think that's a good idea.

Author: Ok. Will add the tech-preview->GA graduation criteria.

#### Dev Preview -> Tech Preview

N.A. This feature will go directly to GA.

#### Tech Preview -> GA

N.A. This feature will go directly to GA.

#### Removing a deprecated feature

### Upgrade / Downgrade Strategy

Upgrade expectations:
* On upgrade, the OVN-K master will create the corresponding ``DNSName`` CRs for each DNS name in the
existing EgressFirewall resources. The ``egressfirewall`` plugin will also start updating the ``.status``
fields of the ``DNSName`` CRs. The scenarios arising out of the order of the update of the various components
are dicussed in [Version Skew Strategy](#version-skew-strategy)

Downgrade expectations:
* On downgrade, the ``DNSName`` CRs may still remain. However, these CRs would not have any impact on how
EgressFirewall ACL rules are implemented in the downgraded cluster. Deleting the CR Definition of ``DNSName``
from the cluster would remove all the ``DNSName`` CRs.

> ### Version Skew Strategy

Reviewer 2: one component has to create the CRD, and the other component has to deal with the possibility that the CRD doesn't exist when it starts up, but is created later

Author: Will the components create the CRD or the CVO?

The following 2 scenarios may occur during the upgrade process:
* Scenario 1: The Cluster DNS operator and the CoreDNS pods are upgraded first and then the OVN-K master pods.

  In this scenario, the ``egressfirewall`` CoreDNS plugin will start inspecting each DNS lookup before the ``DNSName``

>   CRs are created by the OVN-K master. The plugin will just respond with the response received from other plugins for

Author: It will be a no-op, as there will be no match for the DNS name. It has to inspect the query to get the DNS name used in the lookup.

  the DNS lookups. As OVN-K master will be continuing the DNS lookups for DNS names with expired TTLs, CoreDNS will
  also be responding with the corresponding IPs and the TTLs.

* Scenario 2: The OVN-K master pods are upgraded first and then the Cluster DNS operator and the CoreDNS pods.

  In this scenario, the OVN-K master will create ``DNSName`` CRs for each unique DNS name used in EgressFirewall rules.
  However, as the Cluster DNS operator and the CoreDNS pods are still not upgraded, CoreDNS pods will not run the
  ``egressfirewall`` plugin. The OVN-K master will still receive the response for the DNS lookup queries it will send

>   to the CoreDNS.

Reviewer 2: ...which means that egressfirewall DNS rules will not work correctly until after CoreDNS is updated

Author: The rules will function as it did before the upgrade as the CoreDNS will still respond to the DNS lookups done by the OVN-K master. The new features like supporting wildcard DNS names and supporting dynamically changing IP addresses will not work. But the previous support of egressfirewall DNS rules will still be there.

Reviewer 2: The previous support included its own solution for dynamically changing IP addresses. Assuming that we remove the old code from ovn-kubernetes, then during the skew period, neither new ovn-k nor old coredns would be handling dynamic IPs.

### Operational Aspects of API Extensions

#### Failure Modes

#### Support Procedures

## Implementation History

> ## Alternatives

Reviewer 2: I think we probably need to fill in this section, because as the discussion goes on, it turns out that (surprise!) this solution, like the others proposed, has some annoying problems too...

Alternatives:
- The existing system (explain how it works, and its pros and cons)
- Tell users to use a SOCKS or HTTP proxy instead (where the proxy does DNS and can be configured to allow or reject arbitrary DNS names). This would work _much_ better for the cases where it works, but it's much less transparent for the clients. But it's _SO_ much better that it might be worth explicitly pointing this out in our docs that admins should consider if this would work for them.
- Actually modify the DNS response rather than just snooping it; if the client requests a denied DNS name, they just get back `REFUSED`. People suggest this idea a lot as being the "obvious" implementation, but it really only makes sense if you are doing "deny *.example.com", which (I have argued), you should not be doing. In order to support it for the "allow *.example.com" case, then that implies that we would have to block all DNS requests that are not specifically allowed, but no one would expect that saying "allow *.example.com" (or even "deny 0.0.0.0/0") would block all other DNS. So I don't think this is really an option. (Also, it has the problem that if the pod has access to DNS via any other means, then they can subvert your policy.)
- Anything else? What do Calico and Cilium do?

Reviewer 2: https://github.com/freedge/dnsflow !?

Author: In Cilium the IP information corresponding to a DNS name is captured through a DNS Proxy which runs on each Cilium agent: https://docs.cilium.io/en/latest/security/policy/language/#dns-based

Author: So, Cilium does both. It re-resolves the DNS names periodically as well as uses a DNS Proxy to get the response of DNS queries to get the IP information corresponding to the DNS names.

Author: > https://github.com/freedge/dnsflow !?

This solution also is kind of similar to that of Cilium. Instead of running a DNS Proxy on each agent, a separate daemonset is deployed which tracks the pods for each specific node. The daemoset also gets the DNS traffic through DNSTap plugin and extracts the response of DNS queries for the pods running on the specific node. It then updates the OVS flows directly of the node.

Reviewer 6: In another comment thread on this PR, danw mentioned the idea of using gRPC instead of CRs, which seems like an alternative worth mentioning.

not be enough to get the IP addresses of all the subdomains of the wildcard DNS name as each subdomain may have a corresponding
`A` or `AAAA` record. Additionally, the DNS lookup of the wildcard DNS name may also fail to fetch any IP if no `A` or
`AAAA` record exists for the DNS name.

### User Stories

* As an OpenShift cluster administrator, I want to add regular DNS names to EgressFirewall rules, so that I can allow
access to them even if the IP addresses associated with the corresponding DNS records change. 
* As an OpenShift cluster administrator, I want to add wildcard DNS names to EgressFirewall rules, so that I can
allow access to all the subdomains belonging to the wildcard DNS names.

### Goals

* Support allowing access to DNS names even if the IP addresses associated with them changes
* Support usage of wildcard DNS names in EgressFirewall rules.

### Non-Goals

* Support additional DNS resolution functionality in the new CoreDNS plugin. The new plugin will only inspect the
response of DNS resolution by the other existing plugins.
* Support denying of DNS lookups in the new CoreDNS plugin. The new plugin will not stop the DNS lookup itself and will not

> respond with a `REFUSED`/`NXDOMAIN` response code if a EgressFirewall rule denies access to the specific DNS name being queried for.

Author: Added that explicitly in the non-goals.

(`EgressFirewallDNSName`) to improve the integration of CoreDNS with EgressFirewall. This proposal takes the OVN Interconnect (OVN-IC)
architecture into consideration where there are (possibly) multiple OVN-K masters and a centralized OVN-K cluster manager. The OVN-K cluster manager will
create a `EgressFirewallDNSName` CR for each unique DNS name (both regular and wildcard DNS names) used in
the EgressFirewall rules. This CR will be used to store the DNS name along with the current IP addresses, the corresponding
TTL, and the last lookup time. The `EgressFirewallDNSName` CR is meant for communication between
CoreDNS and OVN-K master(s).

The new plugin will inspect each DNS lookup and the corresponding response for the DNS lookup from other
plugins. If the DNS name in the query matches any `EgressFirewallDNSName` CR(s) (regular or wildcard or both), then the
plugin will update the `.status` of the matching `EgressFirewallDNSName` CR(s) with the DNS name along with the IP addresses,
the corresponding TTL, and the last lookup time. The OVN-K master(s) will watch the `EgressFirewallDNSName`

> CRs. Whenever the IP addresses are updated for a `EgressFirewallDNSName` CR, the OVN-K master(s) will update the underlying `AddressSet`

Author: It will directly use the IPs of the DNS name from the updated status of the CR and replace the existing IPs in the underlying AddressSet.

referenced by the ACL rule(s) for the corresponding EgressFirewall rule(s).

A new controller (`EgressFirewallDNSName` controller) will keep track of the next lookup time (TTL + last lookup time)
for each regular DNS name and send a DNS lookup
query to CoreDNS when the minimum TTL expires. However, for a wildcard DNS name a DNS lookup cannot be only performed
on the DNS name as it will not return the IP addresses of all the subdomains. The DNS lookup of the wildcard DNS name may fail to return
any IP address as well. If the lookup for the wildcard DNS name fails, then it will retried using the default TTL (30 minutes). If the lookup
succeeds then the details will be added to the `.status` of the corresponding CR. Thus, the lookups will be performed on the DNS names which

> are updated in the `.status` of the corresponding wildcard `EgressFirewallDNSName` CRs.

Author: The new Controller will remove the stale DNS names from the status. I have explained the process here: https://github.com/openshift/enhancements/pull/1335/files/88cd4ddfd9549b3045f10252e4aeda9199f8bc55#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R276-R282

The following `EgressFirewallDNSName` CRD will be added to the `dns.openshift.io` api-group.

````go
// EgressFirewallDNSName describes a DNS name used in a EgressFirewall rule. It is TechPreviewNoUpgrade only.
type EgressFirewallDNSName struct {
	metav1.TypeMeta `json:",inline"`

	// metadata is the standard object's metadata.
	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// spec is the specification of the desired behavior of the EgressFirewallDNSName.
	Spec EgressFirewallDNSNameSpec `json:"spec,omitempty"`
	// status is the most recently observed status of the EgressFirewallDNSName.
	Status EgressFirewallDNSNameStatus `json:"status,omitempty"`
}

// EgressFirewallDNSNameSpec is a desired state description of EgressFirewallDNSName.
type EgressFirewallDNSNameSpec struct {

> 	// name is the DNS name used in a EgressFirewall rule.

Author: The details of the `DNSName` field of EgressFirewall can be found here: https://github.com/openshift/enhancements/pull/1335/files/88cd4ddfd9549b3045f10252e4aeda9199f8bc55#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R249-R254

The same DNS name can be used in multiple EgressFirewall objects, but only a single EgressFirewallDNSName object will be created for the DNS name.

	// +kubebuilder:validation:Required

> 	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$

Author: The pattern is suggested to be changed here: https://github.com/openshift/enhancements/pull/1335/files/88cd4ddfd9549b3045f10252e4aeda9199f8bc55#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R253

This pattern matches the one mentioned for the `name` field. Only the trailing `.` is optional for the `DNSName` field of EgressFirewall whereas it is mandatory here.

> 	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$

Reviewer: You have to make sure none of the dns "labels" start with "-".  We use this pattern for the labels: https://github.com/openshift/library-go/blob/master/vendor/k8s.io/apimachinery/pkg/util/validation/validation.go#L178

> 	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$

Author: Thanks for pointing it out. I'll update the regex.

	Name string `json:"name"`
}

// EgressFirewallDNSNameStatus defines the observed status of EgressFirewallDNSName.

> type EgressFirewallDNSNameStatus struct {

Author: I will be adding the retry counter field to indicate error while getting the IPs of a DNS name. The retry counter field will be used to remove stale DNS entries from the status.

> type EgressFirewallDNSNameStatus struct {

Reviewer 3: A number doesn't tell the full story however, conditions give you the option to provide much more information, textually. Would it not be good to surface the error messages or some know messages when certain conditions are taking place? This is pretty common across kube land

> type EgressFirewallDNSNameStatus struct {

Author: Should the conditions be for each `resolvedNames` field? As each DNS lookup of each resolved name is independent and having condition for each of them would be easier to maintain.

> type EgressFirewallDNSNameStatus struct {

Reviewer 3: Yes, I think that would make sense

	// The list of matching DNS names and their corresponding IP addresses along with TTL and last
	// DNS lookup time.
	ResolvedNames []EgressFirewallDNSNameStatusItem `json:"resolvedNames,omitempty"`
}

// EgressFirewallDNSNameStatusItem describes the details of a resolved DNS name.
type EgressFirewallDNSNameStatusItem struct {
	// The resolved DNS name corresponding to the Name field of EgressFirewallDNSNameSpec.
	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$
	DNSName string `json:"dnsName"`
	// The IP addresses associated with the DNS name used in a EgressFirewall rule.

> 	IPs []string `json:"ips"`

Author: Makes sense. Will add it.

	// Minimum time-to-live value among all the IP addresses.

> 	TTL int64 `json:"ttl"`

Author: > Should this be required?

The TTL+LastLookupTime will give the next time to lookup for the DNS name. It is required for the DNS name re-resolution performed by the new Controller.

> Do you think you really need an int64? Perhaps an int32 would do?

There is an existing CRD `DNSRecord` which uses `int64` for TTL. It was suggested here:
https://github.com/openshift/enhancements/pull/1335#discussion_r1156647749
The [format](https://github.com/openshift/coredns/blob/master/vendor/github.com/miekg/dns/dns.go#L66-L73) used by CoreDNS is `uint32`.  Can we use `int32` in the API for `uint32` value?

> What is this TTL unit? Is this in seconds? Have you considered using a duration string?
Yes. The TTL is in seconds. I had initially proposed it as a string field. In the latest update I have changed it to `int64` field.

> 	TTL int64 `json:"ttl"`

Reviewer 3: Ok, I think a `int32` is acceptable, but we should call the field `ttlSeconds` in that case

	// Timestamp when the last DNS lookup was successfully completed.

> 	LastLookupTime metav1.Time `json:"lastLookupTime"`

Author: The TTL+LastLookupTime will give the next time to lookup for the DNS name. It is required for the DNS name re-resolution performed by the new Controller.

> 	LastLookupTime metav1.Time `json:"lastLookupTime"`

Reviewer 3: I think in that case it needs a required tag

}
````

### Workflow Description

The workflows for Create, Delete and Update events for EgressFirewall related to DNS names are explained in this section. The workflow
for those events is shown in the following diagram with an example `EgressFirewallDNSName` CR:

> ![Workflow](./coredns-egressfirewall-integration.png)

Reviewer 4: I can't figure out which field `Egress Firewall DNS Name Controller` actually updates?

Author: The Controller will remove stale DNS name entries from the status of the CRs. I have explained it here: https://github.com/openshift/enhancements/pull/1335/files#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R276-R282

#### Create/Update of DNS name

* An OpenShift cluster administrator creates/updates an EgressFirewall resource for a namespace and adds rule(s) containing
DNS name(s).
* The OVN-K cluster manager will create corresponding `EgressFirewallDNSName` CRs for each of the DNS names in the EgressFirewall rules, if not

> already created. Each CR will be created in the `openshift-ovn-kubernetes` namespace. The name of the CR will be assigned using a hash

Reviewer 3: Crossing namespaces means you can't use the garbage collector built into K8s, why do we want all DNSNames in one namespace rather than keeping them in the same namespaces as the EgressFirewall that they belong to?

What data will form the input for the hash? Is this designed to deduplicate entries?

Author: > Crossing namespaces means you can't use the garbage collector built into K8s, why do we want all DNSNames in one namespace rather than keeping them in the same namespaces as the EgressFirewall that they belong to?

The DNS names are not mapped in an 1-1 fashion to the EgressFirewall objects. Same DNS name can be used in multiple EgressFirewall objects in different namespaces, however only a single EgressFirewallDNSName oject should be created for the DNS name. That's why it is created only in `openshift-ovn-kubernetes` namespace.

> What data will form the input for the hash? Is this designed to deduplicate entries?

The DNS name itself is used as the input for the hash as only a single CR should be created for each unique DNS name.

Reviewer 3: This context isn't clear from the enhancement, please make sure that makes it to the document

function (similar to the ComputeHash [here](https://github.com/openshift/kubernetes/blob/master/pkg/controller/controller_utils.go#L1157-L1172))

> prefixed by `dns-`. The `.spec.name` field of the CR will be set to the DNS name (along with a trailing `.`).

Author: DNS servers store DNS names with a trailing period and whenever a DNS request is sent to CoreDNS, the trailing period is added, if it's not already added. So, DNS name matching during lookup will be just simple string matching. 

Additionally, when the status will be updated by CoreDNS, it will add the DNS names with the trailing period. So, matching the DNS names in the status also will be simple string matching.

> prefixed by `dns-`. The `.spec.name` field of the CR will be set to the DNS name (along with a trailing `.`).

Reviewer 3: This context would be good to include here, I didn't understand why this was being done either.
From a UX perspective though, I would expect the values to match between the user input rule and the name here, why not do the trailing dot addition inside the plugin itself?

> prefixed by `dns-`. The `.spec.name` field of the CR will be set to the DNS name (along with a trailing `.`).

Author: The CR is not for the users actually. The whole lifecycle is through OCP components without any user intervention.

Also, as only a single EgressFirewallDNSName CR will be created for an unique DNS name, even if the DNS name is used in multiple EgressFirewall objects, there is no guarantee how the user adds the DNS name (with or without trailing period) in each of the EgressFirewall DNS rules.

Thus, to make it consistent, and also for ease of matching, the `spec.name` will be set with a trailing period.

#### Cluster DNS Operator

Cluster DNS Operator will deploy CoreDNS with the `egressfirewall` plugin enabled by adding it to the corefile. For the EgressFirewall rules to
apply consistently, even for DNS names that are resolved by custom upstreams, it will be added to all server blocks in the corefile. As the
plugin will watch and update the `EgressFirewallDNSName` CRs in the `dns.openshift.io` api-group, proper RBAC permissions will be needed
to be added to the `ClusterRole` for CoreDNS.

The new `EgressFirewallDNSName` controller will be added to the Cluster DNS Operator. The controller will watch the `EgressFirewallDNSName` CRs,
and will send DNS lookup requests for the `spec.name` field. It will also re-resolve the `status.resolvedNames[*].dnsName` fields based on the
corresponding next lookup time(TTL + last lookup time).

For wildcard DNS names, the controller will query for the DNS names that get added to the `.status` of the corresponding `EgressFirewallDNSName` CR,
including the wildcard DNS name, even if it doesn't get added to the `.status`. However, the list of the DNS names to lookup for a wildcard DNS name

> should also not become stale if a DNS name belonging its subdomain is removed. To achieve this a retry counter will be used for the DNS name

Author: I missed adding the retry counter to the API. Will add it.

[here](https://github.com/openshift/kubernetes/blob/master/pkg/controller/controller_utils.go#L1157-L1172))
prefixed by `dns-`. It will also delete a `EgressFirewallDNSName` CR, when all the rules containing the corresponding DNS name are deleted.

#### OVN-K master(s)

The OVN-K master(s) will watch the `EgressFirewallDNSName` CRs. Whenever the `.status` of the CRs will be updated with new IP addresses and corresponding
TTL information for a DNS name, OVN-K master(s) will update the `AddressSet` mapped to the DNS name. This `AddressSet` will be linked
to the ACL rule(s) for the EgressFirewall rule(s) in which the DNS name is used. This will ensure that the latest IP addresses are always updated in
the `AddressSets`.

### Risks and Mitigations

* The `EgressFirewallDNSName` CR will be created by OVN-K Cluster manager whenever a new DNS name is used in a EgressFirewall rule. The CR will be deleted
when the corresponding DNS name is not used in any of the EgressFirewall rules. The `EgressFirewallDNSName` CR should not be modified (created or deleted or
updated) by an user. Doing so may lead to undesired behavior of EgressFirewall.

> ### Drawbacks

Reviewer 4: maybe we can add a note here about extra delay for requests matching wildcarded DNS names, that don't have a wildcarded DNS record?

	//   - False if any of those conditions are unsatisfied.
	// +optional
	// +listType=map
	// +listMapKey=type
	Conditions []metav1.Condition `json:"conditions,omitempty"`
}
````

### Workflow Description

The workflows for the different components related to DNSNameResolver events are explained in this section. The workflow
for those events is shown in the following diagram with an example `DNSNameResolver` CR:

![Workflow](./coredns-egressfirewall-integration.png)

#### Workflow of OVN-K cluster manager

The following pseudocode explains the workflow of the OVN-K cluster manager:
```shell
watch(EgressFirewall)

if match(event.type, "create"):

	for each rule in obj.egress:
		if rule.to.dnsName != ""  && check_dns_name_resolver_exists(rule.to.dnsName) == false:
			create_dns_name_resolver(rule.to.dnsName)

else if match(event.type, "delete"):

	for each rule in obj.egress:
		if rule.to.dnsName != "" && check_dns_name_used_in_any_egress_firewall(rule.to.dnsName) == false:
			delete_dns_name_resolver(rule.to.dnsName)
```

* The OVN-K cluster manager will watch for events related to `EgressFirewall` objects. (Note: Currently, on update of an `EgressFirewall` object, it is

> first deleted and then created again by the OVN-K master(s). Thus, only the create and delete events are handled here.)

Author: I wanted to get some clarifications on this.

- Why is an update event handled by deleting and creating the EgressFirewall object?
- Since cluster manager is also part of ovn-k, wouldn't it be better that the masters and the cluster manager handle the EgressFirewall objects similarly?

From the point of view of cluster manager, if it were to handle the update event, then the cluster manager will have to do the following:

1. Spend CPU cycles to process the EgressFirewall object to get the changes in the DNS names and create/delete DNSNameResolver objects based on the derived changes in the EgressFirewall object
2. Then get a delete event, process the object again and delete DNSNameResolver objects, if required
3. Then get a create event, process the object again and create DNSNameResolver objects, if required

The time and resources spent on processing the update event will not be useful as the same changes will again be brought in by the following delete and create events.

> first deleted and then created again by the OVN-K master(s). Thus, only the create and delete events are handled here.)

Reviewer 4: Maybe I am explaining this in a confusing way, but there will be no extra delete/create event, so steps 2 and 3 won't happen. Inside ovn-k master there will also only be an update event, that will be handled by deleting and re-creating ACLs instead of evaluating the diff and only applying the real changes. Does it make sense?

> first deleted and then created again by the OVN-K master(s). Thus, only the create and delete events are handled here.)

Author: Ok. Got it. I will update the workflow.

> first deleted and then created again by the OVN-K master(s). Thus, only the create and delete events are handled here.)

Author: Added the details for update event.

* If the DNS lookup is for a wildcard DNS name and it matches with the `DNSNameResolver` CR corresponding to the wildcard DNS name,
then the `ocp_dnsnameresolver` plugin will update the `.status` of the corresponding `DNSNameResolver` CR with the wildcard DNS name
and the corresponding current IP addresses along with the TTL and the current lookup time as the last lookup time. If the wildcard DNS name's
corresponding IP addresses and next lookup time (TTL + last lookup time) matches that of any other regular DNS name, added to the `.status`
of the `DNSNameResolver` CR, then the details of the regular DNS name will be removed.
* However, the updates will take place if there is a change in the next lookup time (TTL + last lookup time) for the DNS name or if the next
lookup time is same for the DNS name, but the current IP addresses are different from the existing IP addresses. For the latter, the current
IP addresses are added to the existing IP addresses of the DNS name. This will take care of the scenario where different CoreDNS pods gets
different subsets of the IP addresses in the response to the DNS lookup of the same DNS name. The next time to lookup will be same for them

> though the IP addresses may differ due to upstream DNS load balancing.

Reviewer 4: let's say I have current status with 
```
ips: [1.1.1.1]
TTLSecond: 10
```
and a new DNS reply for the same DNSName with `ip=1.1.1.2, TTL=600`, if I got that right, the status will be updated to
```
ips: [1.1.1.1, 1.1.1.2]
TTLSecond: 600
```
but that TTL is not correct for the first ip, and let's consider the first ip is actually not used anymore, so it needs to be deleted in 10 seconds. How will this work?

Author: The update will take place in 2 scenarios:

- If the next lookup times are different (TTL + last lookup time)
  For example, if the status is:
  ```
  ips: [1.1.1.1]
  ttlSecond: 10
  lastLookupTime: 12:00
  ```
  And a new lookup happens at 12:10 with `ips=1.1.1.2, ttl=60`, then the next lookup time is different for the IP addresses. So the status will be updated to:
  ```
  ips: [1.1.1.2]
  ttlSecond: 60
  lastLookupTime: 12:10
  ```
- If the next lookup time are same but IPs are different
  If we again have the same status:
  ```
  ips: [1.1.1.1]
  ttlSecond: 10
  lastLookupTime: 12:00
  ```
  And a new lookup happens at 12:05 with `ips=1.1.1.2, ttl=5`, then the next lookup time for both the IP addresses are same (12.10). So the status will be updated to:
  ```
  ips: [1.1.1.1, 1.1.1.2]
  ttlSecond: 5
  lastLookupTime: 12:05
  ```

Reviewer 4: not sure I got the first part, so if a new lookup happens at 12:05 instead of 12:10, what would be the result?
Another example I would like to understand is what happens if the status is
```
ips: [1.1.1.1]
ttlSecond: 60
lastLookupTime: 12:00
```
and a new lookup happens at 12:10 with TTL=10 and a new ip 1.1.1.2?
I guess some general algorithm for this merging would be nice to have.

Author: It will match the 1st scenario, where the next time to lookup is different. The status in both the cases that you have mentioned will be updated with the latest IP and TTL. So for the example that I had given:
```
ips: [1.1.1.2]
ttlSecond: 60
lastLookupTime: 12:05
```
For the example that you had given:
```
ips: [1.1.1.2]
ttlSecond: 10
lastLookupTime: 12:10
```
In the scenario that different CoreDNS pods get different sets of IPs, the TTL of those IP addresses will be the same. That's what is being captured in the second scenario mentioned in my previous comment. Thus if the next time to lookup is same for the IPs obtained by different CoreDNS pods, then the IPs will be combined.

However, if the next time to lookup is different, it means that new IPs have been updated in the DNS servers and the existing IPs and TTL have to updated in the status. This is what happens in the 2 examples that you asked. The  next time to lookup does not match and thus the status is updated with the new IP and TTL.

I have already added the details in the pseudocode here (https://github.com/openshift/enhancements/pull/1335/files/6854ffc27aaa56a0b48ba0d01a8a5cb59a151ffe#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R361-R373) and explained it here (https://github.com/openshift/enhancements/pull/1335/files#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R409-R413).

Reviewer 4: thanks for the links, I knew there was some explanation. I have some concerns though:
1. the concept of the "same time": we only have time in seconds, which means rounding errors will happen
2. "In the scenario that different CoreDNS pods get different sets of IPs, the TTL of those IP addresses will be the same" how do you know that? Is it guaranteed by some standard? I think I've seen examples of when this is not the case, and dns responses may have both different ips and next lookup time (now+TTL)

Author: > 1. the concept of the "same time": we only have time in seconds, which means rounding errors will happen

The time matching issue is explained here: https://github.com/openshift/enhancements/pull/1335/files#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R414-R416

> 2\. "In the scenario that different CoreDNS pods get different sets of IPs, the TTL of those IP addresses will be the same" how do you know that? Is it guaranteed by some standard? I think I've seen examples of when this is not the case, and dns responses may have both different ips and next lookup time (now+TTL)

The RFC (https://datatracker.ietf.org/doc/html/rfc2181#section-5.2) clarifies that the TTL of a record will be same for all IPs. Thus, whenever a partial set of IPs are received by different CoreDNS pods, the TTL values will be same.

Reviewer 4: So I have tested this with one of the dns names we had problems with before (`sts.amazonaws.com`) and I could get 2 different replies with 2 different TTLs if I resolve this name from the local machine and the cluster node. You can probably reproduce it the same way, I am not sure if that means that DNS server is not working correctly or if these record are not considered the same RRSet

Author: Different nameservers may respond with different IPs (and TTL) for the same DNS name. The nameservers will be configured in the `/etc/resolv.conf` file. As the CoreDNS pods in a cluster are configured with the same nameserver, there won't be any disparity in the information. You can check how resolution of the external DNS names is configured in CoreDNS: https://github.com/openshift/cluster-dns-operator/blob/master/README.md#how-it-works

Reviewer 4: I think I was told before that there is no guarantee that all nodes in the cluster are configured with the same `resolv.conf` (like it is the config by default, but it can be changed, and that was a part of the problem).
Also, I am not sure the doc you linked says that all `resolv.conf` must be the same on all cluster nodes?

Author: > think I was told before that there is no guarantee that all nodes in the cluster are configured with the same `resolv.conf` (like it is the config by default, but it can be changed, and that was a part of the problem).

We can add in the documentation that if all the nodes are not configured with the same nameserver in the `/etc/resolv.conf` then there are chances that the EgressFirewall won't work correctly. It is already documented in current EgressFirewall documentation that pods should use the same local nameserver as that of the egress firewall controller: https://docs.openshift.com/container-platform/4.12/networking/ovn_kubernetes_network_provider/configuring-egress-firewall-ovn.html#domain-name-server-resolution_configuring-egress-firewall-ovn

Author: I had a discussion with @Miciah last week. Here is a summary of the discussion:
- Even though the nameservers are same on all nodes, if the cluster is geo-distributed, the nameservers may use different IPs in different regions. The assumption that the TTLs will be same may not hold.
- Removing IPs immediately after the expiration of the TTL may cause existing connections to drop. We can wait for a grace period (say 1s) after the TTL expiration to remove an IP.
- When the `DNSNameResolver` sends the request for DNS lookup it can send it to a maximum of 5 random CoreDNS pods rather than the CoreDNS service. This may ensure that the IPs for a DNS name are from different nameservers/regions.

I will update the EP based on this discussion.

Reviewer 6: > The RFC (https://datatracker.ietf.org/doc/html/rfc2181#section-5.2) clarifies that the TTL of a record will be same for all IPs. Thus, whenever a partial set of IPs are received by different CoreDNS pods, the TTL values will be same.

I think the RFE is saying that the TTLs will be the same _at any given point in time_.  However, you will usually see TTLs decrement over time between queries.  Below is a quick example.  Are you accounting for that?  
<details>

Example:
```console
% repeat 3; do dig @8.8.8.8 www.google.com ; sleep 3; done

; <<>> DiG 9.11.36-RedHat-9.11.36-8.el8 <<>> @8.8.8.8 www.google.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 8371
;; flags: qr rd ra; QUERY: 1, ANSWER: 6, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;www.google.com.                        IN      A

;; ANSWER SECTION:
www.google.com.         108     IN      A       64.233.176.105
www.google.com.         108     IN      A       64.233.176.103
www.google.com.         108     IN      A       64.233.176.147
www.google.com.         108     IN      A       64.233.176.104
www.google.com.         108     IN      A       64.233.176.106
www.google.com.         108     IN      A       64.233.176.99

;; Query time: 64 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Thu Aug 03 04:43:09 EDT 2023
;; MSG SIZE  rcvd: 139

; <<>> DiG 9.11.36-RedHat-9.11.36-8.el8 <<>> @8.8.8.8 www.google.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 43282
;; flags: qr rd ra; QUERY: 1, ANSWER: 6, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;www.google.com.                        IN      A

;; ANSWER SECTION:
www.google.com.         105     IN      A       64.233.176.105
www.google.com.         105     IN      A       64.233.176.103
www.google.com.         105     IN      A       64.233.176.147
www.google.com.         105     IN      A       64.233.176.104
www.google.com.         105     IN      A       64.233.176.106
www.google.com.         105     IN      A       64.233.176.99

;; Query time: 34 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Thu Aug 03 04:43:12 EDT 2023
;; MSG SIZE  rcvd: 139

; <<>> DiG 9.11.36-RedHat-9.11.36-8.el8 <<>> @8.8.8.8 www.google.com
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 33288
;; flags: qr rd ra; QUERY: 1, ANSWER: 6, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;www.google.com.                        IN      A

;; ANSWER SECTION:
www.google.com.         102     IN      A       64.233.176.105
www.google.com.         102     IN      A       64.233.176.103
www.google.com.         102     IN      A       64.233.176.147
www.google.com.         102     IN      A       64.233.176.104
www.google.com.         102     IN      A       64.233.176.106
www.google.com.         102     IN      A       64.233.176.99

;; Query time: 28 msec
;; SERVER: 8.8.8.8#53(8.8.8.8)
;; WHEN: Thu Aug 03 04:43:15 EDT 2023
;; MSG SIZE  rcvd: 139

% 
```
</details>

Author: The next lookup time (TTL + last lookup time) is used to check if the expiration time of the IP addresses match or not.

---
title: coredns-egressfirewall-integration
authors:
  - '@arkadeepsen'
reviewers:
  - '@Miciah'
  - '@danwinship'
  - '@JoelSpeed'
  - '@TrilokGeer'

>   - '@jerpeter1'

Reviewer 3: It can be helpful to explain the expertise each reviewer is bringing, what part of this do you want their input on

TTL, and the last lookup time. The `DNSNameResolver` CR is meant for communication between
CoreDNS and OVN-K master(s).

The new plugin will inspect each DNS lookup and the corresponding response for the DNS lookup from other
plugins. If the DNS name in the query matches any `DNSNameResolver` CR(s) (regular or wildcard or both), then the
plugin will update the `.status` of the matching `DNSNameResolver` CR(s) with the DNS name along with the IP addresses,
the corresponding TTL, and the last lookup time. The OVN-K master(s) will watch the `DNSNameResolver`
CRs. Whenever the IP addresses are updated for a `DNSNameResolver` CR, the OVN-K master(s) will update the underlying `AddressSet`
referenced by the ACL rule(s) for the corresponding EgressFirewall rule(s).

A new controller (`DNSNameResolver` controller) will keep track of the next lookup time (TTL + last lookup time)
for each regular DNS name and send a DNS lookup
query to CoreDNS when the minimum TTL expires. However, for a wildcard DNS name a DNS lookup cannot be only performed
on the DNS name as it will not return the IP addresses of all the subdomains. The DNS lookup of the wildcard DNS name may fail to return

> any IP address as well. If the lookup for the wildcard DNS name fails, then it will retried using the default TTL (30 minutes). If the lookup

Reviewer 3: IIRC we added a retry counter, is it documented somewhere? Perhaps a link here to it would be helpful
```
When a lookup fails consistently over time, the record will be pruned as described in ...
```
WDYT?

Reviewer 3: Not yet resolved?

Author: Missed it. Added the link now.

succeeds then the details will be added to the `.status` of the corresponding CR. Thus, the lookups will be performed on the DNS names which
are updated in the `.status` of the corresponding wildcard `DNSNameResolver` CRs.

The following `DNSNameResolver` CRD will be added to the `network.openshift.io` api-group.

> ````go

Reviewer 3: Will do an in-depth review on the API PR now that's up to date, it can be copied back here

	// +patchMergeKey=dnsName
	// +patchStrategy=merge
	// +optional
	ResolvedNames []DNSNameResolverStatusItem `json:"resolvedNames,omitempty" patchStrategy:"merge" patchMergeKey:"dnsName"`
}

// DNSNameResolverStatusItem describes the details of a resolved DNS name.
type DNSNameResolverStatusItem struct {
	// dnsName is the resolved DNS name matching the name field of DNSNameResolverSpec.
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$
	DNSName string `json:"dnsName"`
	// ips contains the list of IP addresses associated with the dnsName.
	// +kubebuilder:validation:Required
	// +listType=set
	IPs []string `json:"ips"`
	// ttlSeconds is the minimum time-to-live value among all the IP addresses.
	// +kubebuilder:validation:Required
	TTLSeconds int32 `json:"ttlSeconds"`
	// lastLookupTime is the timestamp when the last DNS lookup was completed.
	// +kubebuilder:validation:Required
	LastLookupTime *metav1.Time `json:"lastLookupTime"`
	// resolutionFailures keeps the count of how many times the DNS resolution failed for the
	// dnsName field. If the DNS resolution succeeds then the field will be set to zero. Upon

> 	// every failure, the value of the field will be incremented by one. Upon reaching a threshold

Reviewer 3: "a threshold value", do we have this defined?

	// value, the details about the DNS name will be removed.
	ResolutionFailures int `json:"resolutionFailures,omitempty"`
	// conditions provide information about the state of the DNS name.
	//
	// These are the supported conditions:
	//
	//   * Degraded
	//   - True if the following conditions are met:
	//     * The last DNS name resolution failed.
	//   - False if any of those conditions are unsatisfied.
	// +optional
	// +listType=map
	// +listMapKey=type
	Conditions []metav1.Condition `json:"conditions,omitempty"`
}
````

### Workflow Description

The workflows for the different components related to DNSNameResolver events are explained in this section. The workflow
for those events is shown in the following diagram with an example `DNSNameResolver` CR:

![Workflow](./coredns-egressfirewall-integration.png)

#### Workflow of OVN-K cluster manager

The following pseudocode explains the workflow of the OVN-K cluster manager:
```shell
watch(EgressFirewall)

> if match(event.type, "create"):

Reviewer 3: Typically Kube controllers don't distinguish between create/update/delete operations and instead just look at the state of the world for any event and reconcile the difference. You aren't guaranteed to see create/delete events so I would expect your update logic to have to handle those cases anyway.

Is the logic planned to be implemented as separate event handlers or will there be reconciliation logic as described above?

Author: OVN-K has a different retry framework which does distinguish between the different events and also retries actions on objects. I am planning to use the same that's why separated them out: https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/retry/obj_retry.go

within a threshold value (say 5 seconds) of the current lookup time, specifically lies in the range defined by
`[current lookup time - threshold duration, current lookup time + threshold duration]`, then they are considered as same.
* The `Degraded` condition of the DNS name will be set to `false` along with the success reason and message.

##### DNS lookup failure

```shell
update_dns_resolver_failure(objName, dnsName, response):
	obj = get_dns_name_resolver(objName)
	for each resolvedName in obj.status.resolvedNames:
		if resolvedName.dnsName == dnsName:
			if resolvedName.resolutionFailures == threshold:
				remove(obj.status.resolvedNames, resolvedName)
			else
				resolvedName.ttl = min_ttl
				resolvedName.lastLookupTime = now()
				resolvedName.resolutionFailures = resolvedName.resolutionFailures + 1
				cond = condition{type: "Degraded", status: true, reason: failure_reason(response), message: failure_message(response)}
				add_condition(resolvedName.conditions, cond)
	
	update_status_dns_name_resolver(obj)
```

* If a DNS name (regular or wildcard) matches a `DNSNameResolver` CR, then the corresponding resolved name field in the `.status` will be updated.

> * If the corresponding `resolutionFailures` field has reached a threshold value (say 5), then the details of the DNS name will be removed from the

Reviewer 3: Can we decide on a threshold as a group so that we can document it in the API? I think this is an important thing for a user to understand via the API documentation

Author: @Miciah @danwinship @npinaeva Can we take a call on the threshold value for `resolutionFailures`.

Author: I have set to to 5, both in the EP and the api godoc.

> * If the corresponding `resolutionFailures` field has reached a threshold value (say 5), then the details of the DNS name will be removed from the

Reviewer 3: Is this valid only for wildcard or also regular lookups?

Author: Valid for both wildcard and regular lookups.

type IPAddressStr string

// DNSNameResolverStatusItem describes the details of a resolved DNS name.
type DNSNameResolverStatusItem struct {
	// conditions provide information about the state of the DNS name.
	// Known .status.conditions.type is: "Degraded"
	// +optional
	// +listType=map
	// +listMapKey=type
	Conditions []metav1.Condition `json:"conditions,omitempty"`
	// dnsName is the resolved DNS name matching the name field of DNSNameResolverSpec.
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$
	// +kubebuilder:validation:MaxLength=254
	DNSName string `json:"dnsName"`
	// info gives the list of associated IP addresses and the corresponding TTL and last
	// lookup time for the dnsName.
	// +kubebuilder:validation:Required
	// +listType=map
	// +listMapKey=ip
	Info []DNSNameResolverInfo `json:"info"`
	// resolutionFailures keeps the count of how many consecutive times the DNS resolution failed
	// for the dnsName. If the DNS resolution succeeds then the field will be set to zero. Upon
	// every failure, the value of the field will be incremented by one. Upon reaching the value

> 	// of 5, the details about the DNS name will be removed.

Author: Makes sense. Will update the godoc.

		ensure_dns_name_added(resolvedName)
	
	if removal_of_ips_required(obj):
		update_dns_resolver(obj)
		return

	remaining_time_till_grace_period, is_expiring = grace_period_expires_after(obj)

	if is_expiring:
		reconcile_event_after(remaining_time_till_grace_period)
		return

else if match(event.type, "delete"):

	if is_wildcard(obj.spec.name) == true
		|| (is_wildcard(obj.spec.name) == false && match_wildcard_dns_name(obj.spec.name) == false):
		delete_dns_name(obj.spec.name)

	if is_wildcard(obj.spec.name) == true:
		for each resolvedName in obj.status.resolvedNames:
			if resolvedName.dnsName != obj.spec.name && match_regular_dns_name(resolvedName.dnsName) == false:
				delete_dns_name(obj.spec.name)
```
* The `DNSNameResolver` controller will run dns resolution of DNS names in the background which will refresh the `.status` field of the `DNSNameResolver`
CRs upon expiration of the validity of the corresponding IP addresses associated to the resolved names. As the CoreDNS pods are configured using the upstream
nameservers present in the node host's `/etc/resolv.conf` file, the upstream nameservers for different CoreDNS pods may serve different IP addresses (possibly

> with different TTLs) for the same DNS name. Thus, while refreshing the information of a DNS name, the DNS lookup requests are sent to a maximum of 5 randomly

Reviewer 4: I am not sure if sending 5 requests for every dns name is reasonable, because most of the DNS names will probably be resolved to the same ip. Maybe we can only do that for dnsNames that already have more than 1 resolved ip in the list?

Author: We can do that. The idea behind sending the request to 5 coreDNS pods was to try and get the different sets of IPs, if any, for a DNS name.

Reviewer 4: sure, I see the intent, but I am concerned about the performance impact, so trying to see if we can optimize it

Reviewer: Doing one lookup should be fine.  How would you resolve things if you got different ip addresses anyway?

Author: The idea came out of a discussion with @Miciah. It is mentioned here: https://github.com/openshift/enhancements/pull/1335#discussion_r1280694158

> Doing one lookup should be fine. How would you resolve things if you got different ip addresses anyway?

The DNS resolution will be done for a DNS name using the minimum TTL of all the IP addresses.

the EgressFirewall rules. The same DNS name may appear in multiple EgressFirewall rules, but only a single `DNSNameResolver` CR object will be used
for a unique DNS name. This CR will be used to store the DNS name along with the current IP addresses, the corresponding
TTL, and the last lookup time. The `DNSNameResolver` CR is meant for communication between
CoreDNS and OVN-K master(s).

The new plugin will inspect each DNS lookup and the corresponding response for the DNS lookup from other
plugins. If the DNS name in the query matches any `DNSNameResolver` CR(s) (regular or wildcard or both), then the
plugin will update the `.status` of the matching `DNSNameResolver` CR(s) with the DNS name along with the IP addresses,
the corresponding TTL, and the last lookup time. The OVN-K master(s) will watch the `DNSNameResolver`
CRs. Whenever the IP addresses are updated for a `DNSNameResolver` CR, the OVN-K master(s) will update the underlying `AddressSet`
referenced by the ACL rule(s) for the corresponding EgressFirewall rule(s).

A new controller (`DNSNameResolver` controller) will keep track of the next lookup time (TTL + last lookup time)
for each regular DNS name and send a DNS lookup

> query to CoreDNS when the minimum TTL expires. However, for a wildcard DNS name a DNS lookup cannot be only performed

Reviewer: What is the expected minimum TTL?  If it is the same as (or longer than) the rate at which EgressFirewall formerly updated the records (30m?) , what improvement is gained?  If it's more frequent, are we going to be peppering CoreDNS with lookup requests that might not need to be updated?  What if someone mistakenly set the TTL to 1s?

Author: > What is the expected minimum TTL?

The minimum TTL will vary for each DNS names. For a DNS name, the associated IP addresses may have different TTLs. The minimum TTL among these will be used to set the next lookup time of the DNS name. 

> If it is the same as (or longer than) the rate at which EgressFirewall formerly updated the records (30m?) , what improvement is gained? If it's more frequent, are we going to be peppering CoreDNS with lookup requests that might not need to be updated?

The 30m interval is used as a default TTL. If the minimum TTL of a DNS name is less than 30m, then that value will be used for the next lookup.

There is almost no change in behavior regarding the frequency of lookups based on the TTL as compared how it is currently done by EgressFirewall. The only changes are that the default TTL of 30m is not used when the minimum TTL value of a DNS name is higher than 30m, and when the minimum TTL of a DNS name is zero, instead of using the default TTL of 30m, a TTL of 5s is used.

> What if someone mistakenly set the TTL to 1s?

Currently, it will be looked up at a frequency of 1s (both in the existing implementation and the proposed design). We can probably use 5s as the minimum TTL for any TTL value less than 5s.

	metav1.TypeMeta `json:",inline"`

	// metadata is the standard object's metadata.
	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
	metav1.ObjectMeta `json:"metadata,omitempty"`

	// spec is the specification of the desired behavior of the DNSNameResolver.
	// +kubebuilder:validation:Required
	Spec DNSNameResolverSpec `json:"spec"`
	// status is the most recently observed status of the DNSNameResolver.
	// +optional
	Status DNSNameResolverStatus `json:"status,omitempty"`
}

// DNSNameResolverSpec is a desired state description of DNSNameResolver.
type DNSNameResolverSpec struct {
	// name is the DNS name for which the DNS name resolution information will be stored.
	// For a regular DNS name, only the DNS name resolution information of the regular DNS
	// name will be stored. For a wildcard DNS name, the DNS name resolution information
	// of all the DNS names, that matches the wildcard DNS name, will be stored.
	// For a wildcard DNS name, the '*' will match only one label. Additionally, only a single
	// '*' can be used at the beginning of the wildcard DNS name. For example, '*.example.com.'

> 	// will match 'sub1.example.com.' but won't match 'sub2.sub1.example.com.'

Reviewer: Will sub1.example.com. match *.sub1.example.com. or only *.example.com ?  Or both?

Author: sub1.example.com will match only *.example.com.

	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$
	// +kubebuilder:validation:MaxLength=254
	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="spec.name is immutable"
	Name string `json:"name"`
}

// DNSNameResolverStatus defines the observed status of DNSNameResolver.
type DNSNameResolverStatus struct {
	// resolvedNames contains a list of matching DNS names and their corresponding IP addresses
	// along with TTL and last DNS lookup time.
	// +listType=map
	// +listMapKey=dnsName
	// +patchMergeKey=dnsName
	// +patchStrategy=merge
	// +optional
	ResolvedNames []DNSNameResolverStatusItem `json:"resolvedNames,omitempty" patchStrategy:"merge" patchMergeKey:"dnsName"`
}

> // +kubebuilder:validation:Pattern=`^(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$|^s*((([0-9A-Fa-f]{1,4}:){7}(:|([0-9A-Fa-f]{1,4})))|(([0-9A-Fa-f]{1,4}:){6}:([0-9A-Fa-f]{1,4})?)|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){0,1}):([0-9A-Fa-f]{1,4})?))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){0,2}):([0-9A-Fa-f]{1,4})?))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){0,3}):([0-9A-Fa-f]{1,4})?))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){0,4}):([0-9A-Fa-f]{1,4})?))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){0,5}):([0-9A-Fa-f]{1,4})?))|(:(:|((:[0-9A-Fa-f]{1,4}){1,7}))))(%.+)?s*$`

Reviewer: The regex is truncated in the display of this file.

Author: This has been updated in the api PR. I will copy paste it here. We have moved away from regex based validation to OpenAPI based IP address validation.

// IPAddressStr is used for validation of an IP address.
type IPAddressStr string

// DNSNameResolverStatusItem describes the details of a resolved DNS name.
type DNSNameResolverStatusItem struct {
	// conditions provide information about the state of the DNS name.
	// Known .status.conditions.type is: "Degraded"
	// +optional
	// +listType=map
	// +listMapKey=type
	Conditions []metav1.Condition `json:"conditions,omitempty"`
	// dnsName is the resolved DNS name matching the name field of DNSNameResolverSpec. This field can
	// store both regular and wildcard DNS names which match the spec.name field. When the spec.name
	// field contains a regular DNS name, this field will store the same regular DNS name after it is
	// successfully resolved. When the spec.name field contains a wildcard DNS name, this field will

> 	// store the regular DNS names which match the wildcard DNS name and have been successfully resolved.

Reviewer 3: This could be worded better, there will be multiple items in the `resolvedNames` list, so each entry would be a new item in the list, rather than a CSV within a single field. This comment makes more sense when you look at the bigger structure

Rather than `this field will`, perhaps `each resolvedName.dnsName will`?

> 	// store the regular DNS names which match the wildcard DNS name and have been successfully resolved.

Author: I will update the wording.

	// If the wildcard DNS name can also be successfully resolved, then this field will store the wildcard
	// DNS name as well.
	// +kubebuilder:validation:Required
	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.$
	// +kubebuilder:validation:MaxLength=254
	DNSName string `json:"dnsName"`
	// resolvedAddresses gives the list of associated IP addresses and their corresponding TTLs and last
	// lookup times for the dnsName.
	// +kubebuilder:validation:Required
	// +listType=map

> 	// +listMapKey=ip

Reviewer: This is not a map, but would be great if it was, and the key was regular DNS name.

Reviewer 3: Maps are not allowed in Kube APIs, and the DNS name wouldn't make sense as the map key here anyway.
The particular field you have commented on is a list of IP addresses belonging to a particular DNS name, see the field above.
The DNS name is the map key of the list that this object makes up

The structure looks like

```
resolvedNames:
- dnsName: foo.example.com <- this field is the map key for resolvedNames
  resolvedAddresses:
  - ip: 1.2.3.4 <- This field is the map key for resolvedAddresses 
    ttlSeconds: 5
    lastLookupTime: <timestamp>
  - ip: 1.2.3.5
    ttlSeconds: 5
    lastLookupTime: <timestamp>
```

	ResolvedAddresses []DNSNameResolverInfo `json:"resolvedAddresses"`
	// resolutionFailures keeps the count of how many consecutive times the DNS resolution failed
	// for the dnsName. If the DNS resolution succeeds then the field will be set to zero. Upon
	// every failure, the value of the field will be incremented by one. The details about the DNS
	// name will be removed, if the value of resolutionFailures reaches 5 and the TTL of all the
	// associated IP addresses have expired.
	ResolutionFailures int32 `json:"resolutionFailures,omitempty"`
}

type DNSNameResolverInfo struct {

> 	// ip is an IP address associated with the dnsName. The validity of the IP address expires after

Reviewer 3: Parent object contains the name, no need for a name on this struct, see https://github.com/openshift/enhancements/pull/1335/files#r1314730073

	// lastLookupTime + ttlSeconds. To refresh the information a DNS lookup will be performed on the
	// expiration of the IP address's validity. If the information is not refreshed then it will be
	// removed after a grace period of 1 second after the expiration of the IP address's validity.
	// +kubebuilder:validation:Required
	IP IPAddressStr `json:"ip"`
	// ttlSeconds is the time-to-live value of the IP address.
	// +kubebuilder:validation:Required
	TTLSeconds int32 `json:"ttlSeconds"`
	// lastLookupTime is the timestamp when the last DNS lookup was completed.
	// +kubebuilder:validation:Required
	LastLookupTime *metav1.Time `json:"lastLookupTime"`
}
````

> ### Workflow Description

Reviewer: I'd like to see the the steps from the point of view of the DNS lookup - which steps are taken when there is an Egress FW, when there isn't, when there is a wildcard, when there isn't, when a DNS name is in multiple Egress FW, and when there is an address not even in Egress FW, what's the flow?

Author: The DNS lookup details is added in the workflow of the plugin: https://github.com/openshift/enhancements/pull/1335/files#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R333-R500

OR updates an EgressFirewall resource for a namespace and deletes rule(s) containing DNS name(s), the cluster manager will receive the delete/update event.
* The cluster manager will then check if the same DNS names are also used in the EgressFirewall rules in other namespaces. If the
DNS names are not used, then the cluster manager will delete the corresponding `DNSNameResolver` CRs.

#### Workflow of `DNSNameResolver` controller

The following pseudocode explains the workflow of the `DNSNameResolver` controller:
```shell
watch(DNSNameResolver)
run_dns_resolution_in_background()

if match(event.type, "create") || match(event.type, "update"):

	if check_dns_name_added(obj.spec.name) == false:
		add_and_resolve_dns_name(obj.spec.name)

	for each resolvedName in obj.status.resolvedNames:
		ensure_dns_name_added(resolvedName)
	
	if removal_of_ips_required(obj):
		update_dns_resolver(obj)
		return

	remaining_time_till_grace_period, is_expiring = grace_period_expires_after(obj)

	if is_expiring:
		reconcile_event_after(remaining_time_till_grace_period)
		return

else if match(event.type, "delete"):

	if is_wildcard(obj.spec.name) == true

> 		|| (is_wildcard(obj.spec.name) == false && match_wildcard_dns_name(obj.spec.name) == false):

Reviewer: What happens if one of the names of a wildcard is subsequently added to the deny list?  E.g. we allowed *.example.com, then someone gained access to badactor.example.com and so we added badactor.example.com to the deny list?

Author: Only allow rules will be supported for DNS based EgressFirewall rules: https://github.com/openshift/enhancements/pull/1335/files#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R36-R40

It was suggested here by @danwinship: https://github.com/openshift/enhancements/pull/1335#discussion_r1101797917

chosen CoreDNS pods to get as many different associated IP addresses for the DNS name as possible.
* The `DNSNameResolver` controller will watch for the events related to `DNSNameResolver` CRs.
* After it receives the create/update events, it will perform DNS lookup for each of the DNS names corresponding to the `DNSNameResolver` CRs,
if the details of the DNS names are not already added.
* The `DNSNameResolver` controller will store the regular DNS name matching the `.spec.name` field of a `DNSNameResolver` CR and the
corresponding current IP addresses along with the the TTL and the last lookup time for each IP address. Using the TTL and last lookup time of all the associated
IP addresses of a DNS name, a next lookup time for the DNS name is also stored. Based on the next lookup time, the controller will perform DNS
lookups to get the latest IP addresses and TTL information. The DNS lookup will be intercepted by the `ocp_dnsnameresolver` plugin and will enforce an
update of the `.status` of the CR.
* The controller will also store the wildcard DNS name which matches the `.spec.name` field of a `DNSNameResolver` CR. If the DNS lookup is

> successful for the wildcard DNS name, then the DNS name will be stored with the corresponding current IP addresses along with the TTL and the last lookup time

Reviewer: Where is the non-wildcard DNS name stored?

Author: It will also be stored by the controller. Two different maps will be used for storing regular and wildcard DNS names.

> for each IP address. If the DNS lookup is not successful, then it will be retried after a default TTL (30 minutes). The `DNSNameResolver` controller will

Reviewer: If it's 30 minutes, is that an improvement over the current 30 minutes polling time of Egress FW?

will be reconciled after the grace period has expired. This is done so that during the next reconcile the IP addresses can be removed from the `.status` of the
`DNSNameResolver` CR.
* On receiving the delete event, the `DNSNameResolver` controller will remove the details stored regarding the `DNSNameResolver` CRs.
* If the deleted event is for a `DNSNameResolver` corresponding to a regular DNS name, the `DNSNameResolver` controller will remove the details of a regular DNS
name, if it doesn't match a `DNSNameResolver` CR corresponding to a  wildcard DNS name.
* If the deleted event is for a `DNSNameResolver` corresponding to a wildcard DNS name, the `DNSNameResolver` controller will remove the details of a wildcard DNS
name. It will also remove the details of the regular DNS names added to the `.status` of the CR, if each of them doesn't match a `DNSNameResolver` CR corresponding
to a  wildcard DNS name.

#### Workflow of `ocp_dnsnameresolver` CoreDNS plugin

The following pseudocodes explain the workflow of the `ocp_dnsnameresolver` CoreDNS plugin:

##### Watch
```shell
watch(DNSNameResolver)

> if match(event.type, "create"):

Reviewer: This watches updates too, right?

Author: Here only the `spec.name` field of the DNSNameResolver CR is being used. This field stores the DNS name and it is immutable. Thus, watching updates events will not be required.

* On receiving the delete event, the `ocp_dnsnameresolver` plugin will remove the details stored regarding the `DNSNameResolver` CRs.

##### DNS lookup interception

```shell
response = get_response_from_plugin_chain(dnsName)

if match_regular_dns_name(dnsName) == false || match_wildcard_dns_name(dnsName) == false:
	return response

if resolution_failed(response) == false:

	ips, ttls = get_ips_and_ttls(response)
	for each ttl in ttls {
		if ttl = 0:
			ttl = min_ttl
	}
	
	if match_regular_dns_name(dnsName) == true:
		objName = get_regular_dns_name_resolver_name(dnsName)
		update_dns_resolver_success(objName, dnsName, ips, ttls, response)
	
	if match_wildcard_dns_name(dnsName) == true:
		objName = get_wildcard_dns_name_resolver_name(dnsName)
		update_dns_resolver_success(objName, dnsName, ips, ttls, response)

else
  	
	if match_regular_dns_name(dnsName) == true:
		objName = get_regular_dns_name_resolver_name(dnsName)
		update_dns_resolver_failure(objName, dnsName, response)
	
	if match_wildcard_dns_name(dnsName) == true:
		objName = get_wildcard_dns_name_resolver_name(dnsName)
		update_dns_resolver_failure(objName, dnsName, response)

return response
```

> * The `ocp_dnsnameresolver` CoreDNS plugin will intercept the request and the response for the DNS lookups from all the pods. If the DNS name matches

Reviewer: Does the plugin need to intercept both the request and response?  Why not just the response?  What if it updates the ttl in the corresponding `.status` for a request but it gets no response?  Shouldn't the ttl update be based on the receipt of the response and not the receipt of the request?

Author: The CoreDNS plugin chain works in this way only. The request will be received first by the plugin then the response. When the request is received no actions are performed. It is sent to the next plugin in the chain. Actions are performed only after receiving the response. To be explicit I mentioned that it will intercept the request and the response of the the DNS lookups.

* If the set of associated IP addresses of any other regular DNS name is a subset of the wildcard DNS name's associated IP addresses and the corresponding
next lookup time (TTL + last lookup time) of each IP address match, then the details of the regular DNS name will be removed.
* However, the updates will take place if there is a change in the next lookup time (TTL + last lookup time) for the existing IP addresses
or if the current IP addresses are different from the existing IP addresses. For the latter, the current IP addresses are added to the existing
IP addresses of the DNS name. This will take care of the scenario where different CoreDNS pods get different subsets of the IP addresses in the
response to the DNS lookup of the same DNS name.
* Additionally, the exact matching of the next lookup time may never be successful. If the existing next lookup time of a DNS name lies
within a threshold value (say 1 second) of the new lookup time, specifically lies in the range defined by
`[new lookup time - threshold duration, new lookup time + threshold duration]`, then they are considered as same.

> * The `Degraded` condition of the DNS name will be set to `false` along with the success reason and message. The `resolutionFailures` field for

Reviewer: Setting a `Degraded` condition on a DNS name seems like overkill, and beyond the scope and nature of this EP.

Author: It was suggested by @JoelSpeed here: https://github.com/openshift/enhancements/pull/1335#discussion_r1260687028.

A DNS name will be removed from the status of the CR if it fails for 5 times. The `Degraded` condition and the related reason and message will provide clarity why the resolution failed. This will also provide information when we would need to debug any future issue related to EgressFirewall.

#### Variation [optional]

### API Extensions

The validation of [`DNSName`](https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/crd/egressfirewall/v1/types.go#L74-L76) field in
`EgressFirewallDestination` will be updated to accept wildcard DNS names as well. It will be updated from `^([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$` which
accepts only regular DNS names to `^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$`.

````go
// EgressFirewallDestination is the endpoint that traffic is either allowed or denied to
type EgressFirewallDestination struct {
	// ..

	// dnsName is the domain name to allow/deny traffic to. If this is set, cidrSelector must be unset.
	// For a wildcard DNS name, the '*' will match only one label. Additionally, only a single '*' can be
	// used at the beginning of the wildcard DNS name. For example, '*.example.com' will match 'sub1.example.com'
	// but won't match 'sub2.sub1.example.com'
	// +kubebuilder:validation:Pattern=^(\*\.)?([A-Za-z0-9-]+\.)*[A-Za-z0-9-]+\.?$
	DNSName string `json:"dnsName,omitempty"`
	// ..
}
````
The details of the `DNSNameResolver` CRD can be found in the [Proposal](#proposal) section.

> ### Implementation Details/Notes/Constraints [optional]

Reviewer: ```suggestion
### Implementation Details/Notes/Constraints
```

The implementation changes needed for the proposed enhancement are documented in this section for each of the components.

#### Cluster DNS Operator

Cluster DNS Operator will deploy CoreDNS with the `ocp_dnsnameresolver` plugin enabled by adding it to the corefile. For the EgressFirewall rules to
apply consistently, even for DNS names that are resolved by custom upstream nameservers, it will be added to all server blocks in the corefile. As the
plugin will watch and update the `DNSNameResolver` CRs in the `network.openshift.io` api-group, proper RBAC permissions will be needed

> to be added to the `ClusterRole` for CoreDNS.

Reviewer: Why not design a way for the EgressFirewall to configure this if it is activated?  If a customer doesn't even use the EgressFirewall, why would we need to make any changes in the plugin chain?

Author: Additionally, the feature will be behind a feature gate. If the feature gate is enabled, it means that the customer intends to use the feature and all the related components will be enabled including the new plugin.

The new `DNSNameResolver` controller will be added to the Cluster DNS Operator. The controller will watch the `DNSNameResolver` CRs,
and will send DNS lookup requests for the `spec.name` field. It will also re-resolve the `status.resolvedNames[*].dnsName` fields based on the

> corresponding next lookup time (TTL + last lookup time).

Reviewer: Will the status of the `DNSNameResolver` impact the status of the Cluster DNS Operator?  Do we have to mark c-d-o as degraded if the `DNSNameResolver` controller is degraded?

Author: > Is there a way to enable/install this controller only when EgressFirewall is in use?

The feature will be behind a feature gate. If the feature gate is enabled, it means that the customer intends to use the feature and all the related components will be enabled including the new controller.

> Will the status of the DNSNameResolver impact the status of the Cluster DNS Operator? Do we have to mark c-d-o as degraded if the DNSNameResolver controller is degraded?

The status of Cluster DNS Operator won't be impacted by the DNSNameResolver controller. The controller will mainly do lookups for the DNS names.  It will also remove IP addresses whos TTL have expired. These operations won't impact the CDO status.

to the ACL rule(s) for the EgressFirewall rule(s) in which the DNS name is used. This will ensure that the latest IP addresses are always updated in
the `AddressSets`.

#### Cluster Network Operator

To add the support of `DNSNameResolver` CRD in OVN-K, add the flag `--enable-dns-name-resolver` to the corresponding OVN-K pods.

### Risks and Mitigations

* The `DNSNameResolver` CR will be created by OVN-K Cluster manager whenever a new DNS name is used in a EgressFirewall rule. The CR will be deleted
when the corresponding DNS name is not used in any of the EgressFirewall rules. The `DNSNameResolver` CR should not be modified (created or deleted or
updated) by a user. Doing so may lead to undesired behavior of EgressFirewall.
* During the upgrade process if the OVN-K cluster manager and master pods are upgraded first and then the Cluster DNS operator and the CoreDNS pods, the
EgressFirewall functionality may break. The details can be found in [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy) section

### Drawbacks

* Whenever there's a change in the IP addresses or the next lookup time (TTL + last lookup time) for a DNS name, the additional step of updating the

> related `DNSNameResolver` CRs will be executed. This will add some delay to the DNS lookup process. However, this will only

Reviewer: Since this feature is required because the EgressFirewall wasn't designed to keep up with the changes in DNS information, I think it is misleading to imply that this doesn't happen very often.  

We frequently get people asking how to troubleshoot why a DNS lookup takes so long, so I do think we need to find out precisely how much this could add for a wildcard domain with a popular suffix.

happen whenever there's a change in the DNS information.
* For wildcard DNS names, it is not possible to pre-emptively find out the matching regular DNS names which have DNS records in the upstream nameservers.
For these regular DNS names the first lookup may originate from a workload pod, in which case the pod may
see an initial delay until the underlying ACL rules are updated with the corresponding IP addresses. However, this will happen only for the first DNS lookup
of a particular DNS name by any pod. Subsequently, it will be handled by the `DNSNameResolver` controller to keep the IP addresses updated.

## Design Details

> ### Open Questions [optional]

Reviewer: We need to find out how this impacts DNS lookups, and generally how this works, when there are multiple Egress Firewalls, possibly with overlapping wildcard DNS names.

### Test Plan

* This enhancement will be tested through e2e tests by adding EgressFirewall rules containing regular DNS names
and wildcard DNS names. The tests will be added to the `openshift/origin` repository.
* Testing the feature where IP addresses are changed may be a little bit tricky as this will probably include
creation of DNS records and then changing the IP addresses of the DNS records through the e2e tests.

### Graduation Criteria

This feature will initially be released as Tech Preview only.

#### Dev Preview -> Tech Preview

N.A. This feature will go directly to Tech Preview.

#### Tech Preview -> GA (Future work)

* Incorporate the feedback received on the Tech Preview version.
* OpenShift documentation is needed to be updated.
* UTs and e2e tests need to cover all the edge case scenarios.

#### Removing a deprecated feature

> ### Upgrade / Downgrade Strategy

Reviewer: Adding/removing the plugin from CoreDNS will be interesting - need tests to make sure it doesn't greatly impact the upgrade/downgrade times.

Author: Will cover the scenario in e2e tests.

Upgrade expectations:
* On upgrade, the OVN-K cluster manager will create the corresponding `DNSNameResolver` CRs for each DNS name in the
existing EgressFirewall resources. The `DNSNameResolver` controller will start the DNS lookups for the `DNSNameResolver` CRs
and the `ocp_dnsnameresolver` plugin will also start updating the `.status` fields of the `DNSNameResolver` CRs. The scenarios arising
out of the order of the update of the various components are discussed in [Version Skew Strategy](#version-skew-strategy)

> Downgrade expectations:

Reviewer: If you update the regex for EgressFirewall DNS names, you may have some issues with downgrade.  Need to cover both regexes.

#### Cluster DNS Operator

Cluster DNS Operator will deploy CoreDNS with the `egressfirewall` plugin enabled by adding it to the corefile. For the EgressFirewall rules to
apply consistently, even for DNS names that are resolved by custom upstreams, it will be added to all server blocks in the corefile. As the
plugin will watch and update the `EgressFirewallDNSName` CRs in the `dns.openshift.io` api-group, proper RBAC permissions will be needed
to be added to the `ClusterRole` for CoreDNS.

> The new `EgressFirewallDNSName` controller will be added to the Cluster DNS Operator. The controller will watch the `EgressFirewallDNSName` CRs,

Reviewer: It would be better if the design only enabled this controller when there was an egress firewall configured.  If customers do not use that functionality, this adds complicated operations that are not required, and risks regressions with current DNS operations.  

Also, at a minimum, we need to see discussion of whether the status of this new controller will impact the status of the cluster-dns-operator overall.  Are there any conditions of the new controller which wouid make cluster-dns-operator degraded?

Author: > It would be better if the design only enabled this controller when there was an egress firewall configured. If customers do not use that functionality, this adds complicated operations that are not required, and risks regressions with current DNS operations.

The feature is currently planned for tech preview. So, the controller will be enabled if the specific feature gate is enabled. Even when it is enabled, the controller will not be receiving any events if egress firewall is not configured.

> Also, at a minimum, we need to see discussion of whether the status of this new controller will impact the status of the cluster-dns-operator overall. Are there any conditions of the new controller which wouid make cluster-dns-operator degraded?

The controller will be watching the new `EgressFirewallDNSName` CRs and will perform re-resolution based on the TTL of the DNS names which are added to the status of the CRs. This won't impact the status of cluster-dns-operator.

#### OVN-K master(s)

The OVN-K master(s) will watch the `EgressFirewallDNSName` CRs. Whenever the `.status` of the CRs will be updated with new IP addresses and corresponding
TTL information for a DNS name, OVN-K master(s) will update the `AddressSet` mapped to the DNS name. This `AddressSet` will be linked
to the ACL rule(s) for the EgressFirewall rule(s) in which the DNS name is used. This will ensure that the latest IP addresses are always updated in
the `AddressSets`.

### Risks and Mitigations

* The `EgressFirewallDNSName` CR will be created by OVN-K Cluster manager whenever a new DNS name is used in a EgressFirewall rule. The CR will be deleted
when the corresponding DNS name is not used in any of the EgressFirewall rules. The `EgressFirewallDNSName` CR should not be modified (created or deleted or
updated) by an user. Doing so may lead to undesired behavior of EgressFirewall.

### Drawbacks

* Whenever there's a change in the IP addresses or the next lookup time (TTL + last lookup time) for a DNS name, the additional step of updating the

> related `EgressFirewallDNSName` CRs will be executed. This will add some delay to the DNS lookup process. However, this will only

Reviewer: Please present an estimate the delay, for one record and for a cumulative total of daily operations.   Could there possibly be a way to design this so there is never an impact if there is no egress firewall?

Author: > Please present an estimate the delay, for one record and for a cumulative total of daily operations.

The main delay component will be due to the call to the apiserver for updating the status of the CR, if there are any updates. Can you suggest how to provide an estimate of the delay for the same?

> Could there possibly be a way to design this so there is never an impact if there is no egress firewall?

The details of how the new plugin will perform the matching is explained here: https://github.com/openshift/enhancements/pull/1335/files#diff-b27f04c5e2673c47f922695835e746ecb2205a9f7800def32cc15c25af521758R294-R298

I have also responded in detail here: https://github.com/openshift/enhancements/pull/1335#discussion_r1156610054

### Upgrade / Downgrade Strategy

Upgrade expectations:

> * On upgrade, the OVN-K cluster manager will create the corresponding `EgressFirewallDNSName` CRs for each DNS name in the

Reviewer: Perhaps an upgrade should delete any of the old CRs left if a downgrade was performed. Or is there a way that old CRs are marked as stale and cleaned up automatically?

Author: The ovn-k cluster manager will be responsible for the creation and deletion of the CRs. If any orphaned CR is found (for which there are no EgressFirewall DNS rules), then it will be deleted.

### Version Skew Strategy

The following 2 scenarios may occur during the upgrade process:
* Scenario 1: The Cluster DNS operator and the CoreDNS pods are upgraded first and then the OVN-K cluster manager and master pods.

  In this scenario, the `egressfirewall` CoreDNS plugin will start inspecting each DNS lookup before the `EgressFirewallDNSName`
  CRs are created by the OVN-K cluster manager. The plugin will just respond with the response received from other plugins for
  the DNS lookups. As OVN-K master will be continuing the DNS lookups for DNS names with expired TTLs, CoreDNS will
  also be responding with the corresponding IP addresses and the TTLs, and the EgressFirewall functionality will still continue to work as before
  the start of the upgrade.

* Scenario 2: The OVN-K cluster manager and master pods are upgraded first and then the Cluster DNS operator and the CoreDNS pods.

  In this scenario, the OVN-K cluster manager will create `EgressFirewallDNSName` CRs for each unique DNS name used in EgressFirewall rules.
  However, as the Cluster DNS operator and the CoreDNS pods are still not upgraded, CoreDNS pods will not run the

>   `egressfirewall` plugin. Thus, the EgressFirewall functionality will be broken in this scenario.

Reviewer: Perhaps this should be listed in the Risks.

### gRPC connection between OVN-K and CoreDNS

Communication between OVN-K master and CoreDNS happens over a gRPC connection rather than the proposed `EgressFirewallDNSName` CR. Whenever there's a DNS
lookup for a DNS Name which is used in an EgressFirewall rule and the IP addresses associated with DNS name changes, then CoreDNS sends this information to the
OVN-K master. After the underlying ACL rules are updated the OVN-K master responds to the same CoreDNS pod with an OK message. Then the CoreDNS
pod responds to the original DNS lookup request.

> 

Reviewer: I couldn't add a comment in the Pros/Cons section of gRPC, but I like the idea of gRPC for this enhancement.  It allows the proper function for egress firewall without interfering with other DNS lookups.

Author: The gRPC based solution will also need a new plugin which will inspect every DNS lookup. Without that it will not be possible to know whether a lookup is requested for a particular DNS. The difference here would be how CoreDNS communicates the information with OVN-K, which is via a gRPC connection. However, with new OVN interconnect architecture where there can be multiple OVN-K masters, this solution will not scale well.
